// ng-lightning/esm5/lib/progressbar/progressbar.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ChangeDetectionStrategy, ElementRef, Renderer2, OnInit, OnChanges } from '@angular/core';
import { HostService } from '../common/host/host.service';
var NglProgressBar = /** @class */ (function () {
    function NglProgressBar(element, renderer, hostService) {
        this.element = element;
        this.renderer = renderer;
        this.hostService = hostService;
        this.renderer.addClass(this.element.nativeElement, 'slds-progress-bar');
        this.renderer.setAttribute(this.element.nativeElement, 'role', 'progressbar');
        this.renderer.setAttribute(this.element.nativeElement, 'aria-valuemin', '0');
        this.renderer.setAttribute(this.element.nativeElement, 'aria-valuemax', '100');
    }
    Object.defineProperty(NglProgressBar.prototype, "value", {
        get: function () {
            return this._value;
        },
        /**
         * The percentage value of the progress bar.
         */
        set: function (value) {
            this._value = Math.max(0, Math.min(value, 100)); // Trap on [0, 100]
            this.renderer.setAttribute(this.element.nativeElement, 'aria-valuenow', "" + this.value);
        },
        enumerable: true,
        configurable: true
    });
    NglProgressBar.prototype.ngOnInit = function () {
        this.setHostClass();
    };
    NglProgressBar.prototype.ngOnChanges = function () {
        this.setHostClass();
    };
    NglProgressBar.prototype.setHostClass = function () {
        var _a;
        this.hostService.updateClass(this.element, (_a = {},
            _a["slds-progress-bar_" + this.size] = !!this.size,
            _a["slds-progress-bar_" + this.variant] = !!this.variant,
            _a));
    };
    NglProgressBar.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: HostService }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], NglProgressBar.prototype, "value", null);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglProgressBar.prototype, "size", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglProgressBar.prototype, "variant", void 0);
    NglProgressBar = __decorate([
        Component({
            selector: 'ngl-progress-bar',
            template: "<span class=\"slds-progress-bar__value\" [style.width.%]=\"value\"><span class=\"slds-assistive-text\">Progress: {{value}}%</span></span>",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [HostService]
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2, HostService])
    ], NglProgressBar);
    return NglProgressBar;
}());
export { NglProgressBar };
