// ng-lightning/esm2015/lib/pick/pick-option.js
import { __decorate, __metadata } from "tslib";
import { Directive, Input, HostListener, ElementRef, Renderer2, OnInit, OnDestroy } from '@angular/core';
import { NglPick } from './pick';
let NglPickOption = class NglPickOption {
    constructor(element, renderer, nglPick) {
        this.element = element;
        this.renderer = renderer;
        this.nglPick = nglPick;
        this._active = false;
    }
    // Use a getter to prevent direct altering
    get active() {
        return this._active;
    }
    set setValue(value) {
        this._value = value;
    }
    pick(evt) {
        if (evt) {
            evt.preventDefault();
        }
        this.nglPick.selectOption(this._value);
    }
    ngOnInit() {
        this._subscription = this.nglPick.values.subscribe(value => {
            this._active = this._isActive(value);
            const activeClass = this.nglPickActiveClass || this.nglPick.nglPickActiveClass;
            if (activeClass) {
                if (this.active) {
                    this.renderer.addClass(this.element.nativeElement, activeClass);
                }
                else {
                    this.renderer.removeClass(this.element.nativeElement, activeClass);
                }
            }
        });
    }
    ngOnDestroy() {
        this._subscription.unsubscribe();
        this.nglPick.optionRemoved(this._value);
    }
    _isActive(value) {
        if (this.nglPick.isMultiple) {
            if (!value) {
                return false;
            }
            return Array.isArray(value) ? value.indexOf(this._value) > -1 : !!value[this._value];
        }
        else {
            return this._value === value;
        }
    }
};
NglPickOption.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: NglPick }
];
__decorate([
    Input('nglPickOption'),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], NglPickOption.prototype, "setValue", null);
__decorate([
    Input(),
    __metadata("design:type", String)
], NglPickOption.prototype, "nglPickActiveClass", void 0);
__decorate([
    HostListener('click'),
    HostListener('keydown.Space', ['$event']),
    HostListener('keydown.Enter', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Event]),
    __metadata("design:returntype", void 0)
], NglPickOption.prototype, "pick", null);
NglPickOption = __decorate([
    Directive({
        selector: '[nglPickOption]',
        exportAs: 'nglPickOption',
        host: {
            'role': 'button',
        },
    }),
    __metadata("design:paramtypes", [ElementRef, Renderer2, NglPick])
], NglPickOption);
export { NglPickOption };
