// ng-lightning/esm2015/lib/pick/pick.js
import { __decorate, __metadata } from "tslib";
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { InputBoolean } from '../util/convert';
let NglPick = class NglPick {
    constructor() {
        this.values = new BehaviorSubject(null);
        this.nglPickChange = new EventEmitter();
        this.nglOptionDestroyed = new EventEmitter();
        this.isMultiple = false;
    }
    set setSelected(selected) {
        this.selected = selected;
        this.ngAfterContentInit();
    }
    ngAfterContentInit() {
        this.values.next(this.selected);
    }
    selectOption(value) {
        let next;
        if (this.isMultiple) {
            if (Array.isArray(this.selected)) {
                // Remove if already there or add to selection
                const index = this.selected.indexOf(value);
                next = index > -1
                    ? [...this.selected.slice(0, index), ...this.selected.slice(index + 1)]
                    : [...this.selected, value];
            }
            else {
                next = Object.assign({}, this.selected, { [value]: !this.selected[value] });
            }
        }
        else {
            next = value;
        }
        this.nglPickChange.emit(next);
    }
    optionRemoved(value) {
        if (this.isMultiple && !this.selected) {
            return;
        }
        let emit;
        if (this.isMultiple) {
            emit = Array.isArray(this.selected) ? this.selected.indexOf(value) > -1 : !!this.selected[value];
        }
        else {
            emit = this.selected === value;
        }
        if (emit) {
            setTimeout(() => this.nglOptionDestroyed.emit(value));
        }
    }
};
__decorate([
    Input('nglPick'),
    __metadata("design:type", String),
    __metadata("design:paramtypes", [String])
], NglPick.prototype, "setSelected", null);
__decorate([
    Input(),
    __metadata("design:type", String)
], NglPick.prototype, "nglPickActiveClass", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NglPick.prototype, "nglPickChange", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NglPick.prototype, "nglOptionDestroyed", void 0);
__decorate([
    Input('nglPickMultiple'), InputBoolean(),
    __metadata("design:type", Object)
], NglPick.prototype, "isMultiple", void 0);
NglPick = __decorate([
    Directive({
        selector: '[nglPick]',
    })
], NglPick);
export { NglPick };
