// ng-lightning/esm5/lib/pick/pick.js
import { __decorate, __metadata, __read, __spread } from "tslib";
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { InputBoolean } from '../util/convert';
var NglPick = /** @class */ (function () {
    function NglPick() {
        this.values = new BehaviorSubject(null);
        this.nglPickChange = new EventEmitter();
        this.nglOptionDestroyed = new EventEmitter();
        this.isMultiple = false;
    }
    Object.defineProperty(NglPick.prototype, "setSelected", {
        set: function (selected) {
            this.selected = selected;
            this.ngAfterContentInit();
        },
        enumerable: true,
        configurable: true
    });
    NglPick.prototype.ngAfterContentInit = function () {
        this.values.next(this.selected);
    };
    NglPick.prototype.selectOption = function (value) {
        var _a;
        var next;
        if (this.isMultiple) {
            if (Array.isArray(this.selected)) {
                // Remove if already there or add to selection
                var index = this.selected.indexOf(value);
                next = index > -1
                    ? __spread(this.selected.slice(0, index), this.selected.slice(index + 1)) : __spread(this.selected, [value]);
            }
            else {
                next = Object.assign({}, this.selected, (_a = {}, _a[value] = !this.selected[value], _a));
            }
        }
        else {
            next = value;
        }
        this.nglPickChange.emit(next);
    };
    NglPick.prototype.optionRemoved = function (value) {
        var _this = this;
        if (this.isMultiple && !this.selected) {
            return;
        }
        var emit;
        if (this.isMultiple) {
            emit = Array.isArray(this.selected) ? this.selected.indexOf(value) > -1 : !!this.selected[value];
        }
        else {
            emit = this.selected === value;
        }
        if (emit) {
            setTimeout(function () { return _this.nglOptionDestroyed.emit(value); });
        }
    };
    __decorate([
        Input('nglPick'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], NglPick.prototype, "setSelected", null);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglPick.prototype, "nglPickActiveClass", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NglPick.prototype, "nglPickChange", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NglPick.prototype, "nglOptionDestroyed", void 0);
    __decorate([
        Input('nglPickMultiple'), InputBoolean(),
        __metadata("design:type", Object)
    ], NglPick.prototype, "isMultiple", void 0);
    NglPick = __decorate([
        Directive({
            selector: '[nglPick]',
        })
    ], NglPick);
    return NglPick;
}());
export { NglPick };
