// ng-lightning/esm5/lib/checkboxes/group/checkbox-group.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ChangeDetectionStrategy, HostBinding, ContentChildren, QueryList } from '@angular/core';
import { toBoolean, InputBoolean } from '../../util/convert';
import { uniqueId } from '../../util/util';
import { NglCheckboxOption } from './checkbox-option';
var NglCheckboxGroup = /** @class */ (function () {
    function NglCheckboxGroup() {
        this.uid = uniqueId('checkbox-group');
        this._type = 'list';
    }
    Object.defineProperty(NglCheckboxGroup.prototype, "hasError", {
        get: function () {
            return toBoolean(this.error);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NglCheckboxGroup.prototype, "errorId", {
        get: function () {
            return "error_" + this.uid;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NglCheckboxGroup.prototype, "type", {
        get: function () {
            return this._type;
        },
        set: function (type) {
            this._type = type;
            this.updateChildrenType();
        },
        enumerable: true,
        configurable: true
    });
    NglCheckboxGroup.prototype.ngOnChanges = function (changes) {
        var _this = this;
        if (changes.error && this.options) {
            this.options.forEach(function (option) {
                option.setError(_this.error ? _this.errorId : null);
            });
        }
    };
    NglCheckboxGroup.prototype.ngAfterContentInit = function () {
        this.updateChildrenType();
    };
    NglCheckboxGroup.prototype.updateChildrenType = function () {
        var _this = this;
        if (!this.options) {
            return;
        }
        this.options.forEach(function (option) {
            option.type = _this.type;
        });
    };
    __decorate([
        ContentChildren(NglCheckboxOption),
        __metadata("design:type", QueryList)
    ], NglCheckboxGroup.prototype, "options", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglCheckboxGroup.prototype, "label", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglCheckboxGroup.prototype, "error", void 0);
    __decorate([
        HostBinding('class.slds-has-error'),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [])
    ], NglCheckboxGroup.prototype, "hasError", null);
    __decorate([
        Input(), InputBoolean(),
        __metadata("design:type", Boolean)
    ], NglCheckboxGroup.prototype, "required", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], NglCheckboxGroup.prototype, "type", null);
    NglCheckboxGroup = __decorate([
        Component({
            selector: 'ngl-checkbox-group,[ngl-checkbox-group]',
            template: "\n<legend class=\"slds-form-element__legend slds-form-element__label\"><abbr class=\"slds-required\" *ngIf=\"required\" title=\"required\">*</abbr><span [nglInternalOutlet]=\"label\"></span></legend>\n<div class=\"slds-form-element__control\">\n  <div class=\"slds-checkbox_button-group\" *ngIf=\"type === 'button'; else contentTpl\">\n    <ng-container *ngTemplateOutlet=\"contentTpl\"></ng-container>\n  </div>\n</div>\n<div class=\"slds-form-element__help\" *ngIf=\"hasError\" [id]=\"errorId\" [nglInternalOutlet]=\"error\"></div>\n<ng-template #contentTpl>\n  <ng-content></ng-content>\n</ng-template>",
            changeDetection: ChangeDetectionStrategy.OnPush,
            host: {
                '[class.slds-form-element]': 'true',
            }
        })
    ], NglCheckboxGroup);
    return NglCheckboxGroup;
}());
export { NglCheckboxGroup };
