// ng-lightning/esm5/lib/checkboxes/group/checkbox-option.js
import { __decorate, __metadata } from "tslib";
import { Component, TemplateRef, Input, ContentChild, ChangeDetectionStrategy, ChangeDetectorRef, ElementRef } from '@angular/core';
import { NglCheckboxInput } from '../input/input';
import { HostService } from '../../common/host/host.service';
var NglCheckboxOption = /** @class */ (function () {
    function NglCheckboxOption(cd, element, hostService) {
        this.cd = cd;
        this.element = element;
        this.hostService = hostService;
    }
    Object.defineProperty(NglCheckboxOption.prototype, "type", {
        get: function () {
            return this._type;
        },
        set: function (type) {
            this._type = type;
            this.setHostClass();
            this.cd.detectChanges();
        },
        enumerable: true,
        configurable: true
    });
    NglCheckboxOption.prototype.setError = function (id) {
        this.input.describedBy = id;
    };
    NglCheckboxOption.prototype.setHostClass = function () {
        var _a;
        this.hostService.updateClass(this.element, (_a = {},
            _a["slds-checkbox"] = this.type === 'list',
            _a["slds-button"] = this.type === 'button',
            _a["slds-checkbox_button"] = this.type === 'button',
            _a));
    };
    NglCheckboxOption.ctorParameters = function () { return [
        { type: ChangeDetectorRef },
        { type: ElementRef },
        { type: HostService }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglCheckboxOption.prototype, "label", void 0);
    __decorate([
        ContentChild(NglCheckboxInput, { static: true }),
        __metadata("design:type", NglCheckboxInput)
    ], NglCheckboxOption.prototype, "input", void 0);
    NglCheckboxOption = __decorate([
        Component({
            selector: 'ngl-checkbox-option',
            template: "\n<ng-content></ng-content>\n<label class=\"slds-checkbox__label\" *ngIf=\"type === 'list'\" [attr.for]=\"input.id\"><span class=\"slds-checkbox_faux\"></span><span class=\"slds-form-element__label\" [nglInternalOutlet]=\"label\"></span></label>\n<label class=\"slds-checkbox_button__label\" *ngIf=\"type === 'button'\" [attr.for]=\"input.id\"><span class=\"slds-checkbox_faux\" [nglInternalOutlet]=\"label\"></span></label>",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [HostService]
        }),
        __metadata("design:paramtypes", [ChangeDetectorRef, ElementRef, HostService])
    ], NglCheckboxOption);
    return NglCheckboxOption;
}());
export { NglCheckboxOption };
