// ng-lightning/esm5/lib/input/input/input.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ChangeDetectionStrategy, ContentChild, TemplateRef, HostBinding, AfterContentInit, ChangeDetectorRef, OnChanges, OnDestroy } from '@angular/core';
import { NglInputElement } from '../element/element';
import { toBoolean, InputBoolean } from '../../util/convert';
var NglInput = /** @class */ (function () {
    function NglInput(cd) {
        this.cd = cd;
    }
    Object.defineProperty(NglInput.prototype, "hasError", {
        get: function () {
            return toBoolean(this.error);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NglInput.prototype, "errorId", {
        get: function () {
            return "error_" + this._uid;
        },
        enumerable: true,
        configurable: true
    });
    NglInput.prototype.ngOnChanges = function () {
        this.input.describedBy = this.error ? this.errorId : null;
    };
    NglInput.prototype.ngAfterContentInit = function () {
        var _this = this;
        if (!this.input) {
            throw Error("[ng-lightning] Couldn't find an <input> with [ngl] attribute inside NglInput");
        }
        this.ɵRequiredSubscription = this.input.ɵRequiredSubject.subscribe(function (required) {
            _this.required = required;
            _this.cd.detectChanges();
        });
        this._uid = this.input.id;
        this.cd.detectChanges();
    };
    NglInput.prototype.ngOnDestroy = function () {
        if (this.ɵRequiredSubscription) {
            this.ɵRequiredSubscription.unsubscribe();
            this.ɵRequiredSubscription = null;
        }
    };
    NglInput.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        ContentChild(NglInputElement, { static: true }),
        __metadata("design:type", NglInputElement)
    ], NglInput.prototype, "input", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglInput.prototype, "label", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglInput.prototype, "error", void 0);
    __decorate([
        Input(), InputBoolean(),
        __metadata("design:type", Boolean)
    ], NglInput.prototype, "stacked", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglInput.prototype, "fieldLevelHelpTooltip", void 0);
    __decorate([
        HostBinding('class.slds-has-error'),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [])
    ], NglInput.prototype, "hasError", null);
    NglInput = __decorate([
        Component({
            selector: 'ngl-input,[ngl-input]',
            template: "\n<label [nglFormLabel]=\"label\" [attr.for]=\"_uid\" [required]=\"required\"></label>\n<ngl-form-help *ngIf=\"fieldLevelHelpTooltip\" [content]=\"fieldLevelHelpTooltip\"></ngl-form-help>\n<div class=\"slds-form-element__control\">\n  <ng-content></ng-content>\n</div>\n<div class=\"slds-form-element__help\" *ngIf=\"hasError\" [id]=\"errorId\" [nglInternalOutlet]=\"error\"></div>",
            changeDetection: ChangeDetectionStrategy.OnPush,
            host: {
                '[class.slds-form-element]': 'true',
            }
        }),
        __metadata("design:paramtypes", [ChangeDetectorRef])
    ], NglInput);
    return NglInput;
}());
export { NglInput };
