// ng-lightning/esm2015/lib/radio-group/radio-option.js
import { __decorate, __metadata } from "tslib";
import { Component, TemplateRef, Input, ContentChild, ChangeDetectionStrategy, ChangeDetectorRef, HostBinding, OnInit, AfterContentInit, OnDestroy } from '@angular/core';
import { NglRadioGroup } from './radio-group';
import { NglRadioInput } from './input/input';
let NglRadioOption = class NglRadioOption {
    constructor(radioGroup, cd) {
        this.radioGroup = radioGroup;
        this.cd = cd;
        this.subscriptions = [];
    }
    get isTypeList() {
        return this.type === 'list';
    }
    get isTypeButton() {
        return this.type === 'button';
    }
    ngOnInit() {
        this.subscriptions.push(this.radioGroup.type$.subscribe((type) => {
            this.type = type;
            this.cd.detectChanges();
        }), this.radioGroup.error$.subscribe((errorId) => {
            this.input.describedBy = errorId;
        }));
    }
    ngAfterContentInit() {
        this.input.name = this.radioGroup.uid;
    }
    ngOnDestroy() {
        this.subscriptions.forEach((s) => s.unsubscribe());
    }
};
NglRadioOption.ctorParameters = () => [
    { type: NglRadioGroup },
    { type: ChangeDetectorRef }
];
__decorate([
    Input(),
    __metadata("design:type", Object)
], NglRadioOption.prototype, "label", void 0);
__decorate([
    ContentChild(NglRadioInput, { static: true }),
    __metadata("design:type", NglRadioInput)
], NglRadioOption.prototype, "input", void 0);
__decorate([
    HostBinding('class.slds-radio'),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], NglRadioOption.prototype, "isTypeList", null);
__decorate([
    HostBinding('class.slds-button'),
    HostBinding('class.slds-radio_button'),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], NglRadioOption.prototype, "isTypeButton", null);
NglRadioOption = __decorate([
    Component({
        selector: 'ngl-radio-option',
        template: "\n<ng-content></ng-content>\n<label class=\"slds-radio__label\" *ngIf=\"type === 'list'\" [attr.for]=\"input.id\"><span class=\"slds-radio_faux\"></span><span class=\"slds-form-element__label\" [nglInternalOutlet]=\"label\"></span></label>\n<label class=\"slds-radio_button__label\" *ngIf=\"type === 'button'\" [attr.for]=\"input.id\"><span class=\"slds-radio_faux\" [nglInternalOutlet]=\"label\"></span></label>",
        changeDetection: ChangeDetectionStrategy.OnPush
    }),
    __metadata("design:paramtypes", [NglRadioGroup, ChangeDetectorRef])
], NglRadioOption);
export { NglRadioOption };
