// ng-lightning/esm2015/lib/files/file.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ChangeDetectionStrategy, ElementRef, Renderer2, HostBinding, TemplateRef } from '@angular/core';
let NglFile = class NglFile {
    constructor(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.renderer.addClass(this.element.nativeElement, 'slds-file');
        this.renderer.addClass(this.element.nativeElement, 'slds-file_card');
    }
};
NglFile.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
__decorate([
    HostBinding('class.slds-has-title'),
    Input(),
    __metadata("design:type", Object)
], NglFile.prototype, "text", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], NglFile.prototype, "iconName", void 0);
NglFile = __decorate([
    Component({
        selector: 'ngl-file',
        changeDetection: ChangeDetectionStrategy.OnPush,
        template: "\n<figure>\n  <ng-content></ng-content>\n  <figcaption class=\"slds-file__title slds-file__title_card\" *ngIf=\"text\">\n    <div class=\"slds-media slds-media_small slds-media_center\">\n      <div class=\"slds-media__figure slds-line-height_reset\" *ngIf=\"iconName\">\n        <ngl-icon [iconName]=\"iconName\"></ngl-icon>\n      </div>\n      <div class=\"slds-media__body\"><span class=\"slds-file__text slds-truncate\" [title]=\"text\" [nglInternalOutlet]=\"text\"></span></div>\n    </div>\n  </figcaption>\n</figure>"
    }),
    __metadata("design:paramtypes", [ElementRef, Renderer2])
], NglFile);
export { NglFile };
