// ng-lightning/esm5/lib/accordion/accordion-item.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, Output, EventEmitter, ChangeDetectionStrategy, ElementRef, Renderer2 } from '@angular/core';
import { NglAccordionSection } from './accordion-section';
import { uniqueId } from '../util/util';
var NglAccordionItem = /** @class */ (function () {
    function NglAccordionItem(element, renderer) {
        this.toggle = new EventEmitter();
        this.uid = uniqueId('accordion-item');
        renderer.addClass(element.nativeElement, 'slds-accordion__list-item');
    }
    NglAccordionItem.prototype.onToggle = function () {
        this.toggle.emit();
    };
    NglAccordionItem.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], NglAccordionItem.prototype, "isActive", void 0);
    __decorate([
        Input(),
        __metadata("design:type", NglAccordionSection)
    ], NglAccordionItem.prototype, "section", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NglAccordionItem.prototype, "toggle", void 0);
    NglAccordionItem = __decorate([
        Component({
            // tslint:disable-next-line:component-selector
            selector: 'li[nglAccordionItem]',
            changeDetection: ChangeDetectionStrategy.OnPush,
            template: "\n<section class=\"slds-accordion__section\" [class.slds-is-open]=\"isActive\">\n  <div class=\"slds-accordion__summary\">\n    <h3 class=\"slds-accordion__summary-heading\" (click)=\"onToggle()\">\n      <button class=\"slds-button slds-button_reset slds-accordion__summary-action\" [attr.aria-controls]=\"uid\" [attr.aria-expanded]=\"isActive\">\n        <svg class=\"slds-accordion__summary-action-icon slds-button__icon slds-button__icon_left\" nglIconName=\"utility:switch\"></svg><span class=\"slds-truncate\" [nglInternalOutlet]=\"section.label\" [nglInternalOutletContext]=\"{$implicit: section.labelContext}\"></span>\n      </button>\n    </h3>\n  </div>\n  <div class=\"slds-accordion__content\" [attr.hidden]=\"isActive ? null : ''\" [id]=\"uid\">\n    <ng-container *ngIf=\"isActive\">\n      <ng-template [ngTemplateOutlet]=\"section.templateRef\"></ng-template>\n    </ng-container>\n  </div>\n</section>"
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2])
    ], NglAccordionItem);
    return NglAccordionItem;
}());
export { NglAccordionItem };
