// ng-lightning/esm5/lib/buttons/button.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ElementRef, Renderer2, ChangeDetectionStrategy, OnInit, OnChanges, SimpleChanges } from '@angular/core';
import { HostService } from '../common/host/host.service';
var NglButton = /** @class */ (function () {
    function NglButton(el, renderer, hostService) {
        this.el = el;
        this.renderer = renderer;
        this.hostService = hostService;
        /**
         * Changes the appearance of the button.
         */
        this.variant = 'neutral';
        /**
         * Describes the position of the icon with respect to ng-content.
         */
        this.iconPosition = 'left';
        this.renderer.addClass(this.el.nativeElement, 'slds-button');
    }
    NglButton.prototype.ngOnInit = function () {
        this.setHostClass();
    };
    NglButton.prototype.ngOnChanges = function (changes) {
        if (changes.variant) {
            this.setHostClass();
        }
    };
    NglButton.prototype.hasLeftIcon = function () {
        return this.iconName && (!this.iconPosition || this.iconPosition === 'left');
    };
    NglButton.prototype.hasRightIcon = function () {
        return this.iconName && this.iconPosition === 'right';
    };
    NglButton.prototype.setHostClass = function () {
        var _a;
        this.hostService.updateClass(this.el, (_a = {},
            _a["slds-button_" + this.variant] = this.variant && this.variant !== 'base',
            _a));
    };
    NglButton.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: HostService }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglButton.prototype, "variant", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglButton.prototype, "iconName", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglButton.prototype, "iconPosition", void 0);
    NglButton = __decorate([
        Component({
            // tslint:disable-next-line:component-selector
            selector: '[nglButton]',
            template: "\n<svg class=\"slds-button__icon slds-button__icon_left\" *ngIf=\"hasLeftIcon()\" [nglIconName]=\"iconName\"></svg>\n<ng-content></ng-content>\n<svg class=\"slds-button__icon slds-button__icon_right\" *ngIf=\"hasRightIcon()\" [nglIconName]=\"iconName\"></svg>",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [HostService]
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2, HostService])
    ], NglButton);
    return NglButton;
}());
export { NglButton };
