// ng-lightning/esm5/lib/buttons/button-states.js
import { __decorate, __extends, __metadata } from "tslib";
import { Component, Input, ElementRef, Renderer2, ChangeDetectionStrategy } from '@angular/core';
var NglButtonStateOn = /** @class */ (function () {
    function NglButtonStateOn(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.renderer.addClass(this.el.nativeElement, this.getHostClass());
    }
    NglButtonStateOn.prototype.getHostClass = function () {
        return 'slds-text-selected';
    };
    NglButtonStateOn.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglButtonStateOn.prototype, "iconName", void 0);
    NglButtonStateOn = __decorate([
        Component({
            selector: 'ngl-state-on',
            template: "\n<svg class=\"slds-button__icon slds-button__icon_small slds-button__icon_left\" *ngIf=\"iconName\" [nglIconName]=\"iconName\"></svg>\n<ng-content></ng-content>",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2])
    ], NglButtonStateOn);
    return NglButtonStateOn;
}());
export { NglButtonStateOn };
var NglButtonStateOff = /** @class */ (function (_super) {
    __extends(NglButtonStateOff, _super);
    function NglButtonStateOff() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NglButtonStateOff.prototype.getHostClass = function () {
        return 'slds-text-not-selected';
    };
    NglButtonStateOff = __decorate([
        Component({
            selector: 'ngl-state-off',
            template: "\n<svg class=\"slds-button__icon slds-button__icon_small slds-button__icon_left\" *ngIf=\"iconName\" [nglIconName]=\"iconName\"></svg>\n<ng-content></ng-content>",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], NglButtonStateOff);
    return NglButtonStateOff;
}(NglButtonStateOn));
export { NglButtonStateOff };
var NglButtonStateHover = /** @class */ (function (_super) {
    __extends(NglButtonStateHover, _super);
    function NglButtonStateHover() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NglButtonStateHover.prototype.getHostClass = function () {
        return 'slds-text-selected-focus';
    };
    NglButtonStateHover = __decorate([
        Component({
            selector: 'ngl-state-hover',
            template: "\n<svg class=\"slds-button__icon slds-button__icon_small slds-button__icon_left\" *ngIf=\"iconName\" [nglIconName]=\"iconName\"></svg>\n<ng-content></ng-content>",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], NglButtonStateHover);
    return NglButtonStateHover;
}(NglButtonStateOn));
export { NglButtonStateHover };
