// ng-lightning/esm5/lib/buttons/button-stateful.js
import { __decorate, __metadata } from "tslib";
import { Directive, Input, Output, EventEmitter, HostListener, ElementRef, Renderer2, OnInit, OnChanges } from '@angular/core';
import { InputBoolean } from '../util/convert';
import { HostService } from '../common/host/host.service';
var NglButtonStateful = /** @class */ (function () {
    function NglButtonStateful(el, renderer, hostService) {
        this.el = el;
        this.renderer = renderer;
        this.hostService = hostService;
        /**
         * Triggered when the button is clicked.
         */
        this.stateChange = new EventEmitter();
        /**
         * Appearance.
         */
        this.variant = 'neutral';
        this.renderer.addClass(this.el.nativeElement, 'slds-button');
        this.renderer.addClass(this.el.nativeElement, 'slds-button_stateful');
        this.renderer.setAttribute(this.el.nativeElement, 'aria-live', 'assertive');
    }
    NglButtonStateful.prototype.onSelectChange = function () {
        this.stateChange.emit(!this.state);
    };
    NglButtonStateful.prototype.onFocusToggle = function (focused) {
        this.focused = !!+focused;
        if (!this.focused) {
            this.setHostClass();
        }
    };
    NglButtonStateful.prototype.ngOnInit = function () {
        this.setHostClass();
    };
    NglButtonStateful.prototype.ngOnChanges = function () {
        this.setHostClass();
    };
    NglButtonStateful.prototype.setHostClass = function () {
        var _a;
        this.hostService.updateClass(this.el, (_a = {},
            _a["slds-button_" + (this.variant === 'text' ? 'reset' : this.variant)] = !!this.variant,
            _a["slds-is-selected-clicked"] = this.state && this.focused,
            _a["slds-is-selected"] = this.state && !this.focused,
            _a["slds-not-selected"] = !this.state,
            _a));
    };
    NglButtonStateful.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: HostService }
    ]; };
    __decorate([
        Input(), InputBoolean(),
        __metadata("design:type", Boolean)
    ], NglButtonStateful.prototype, "state", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NglButtonStateful.prototype, "stateChange", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglButtonStateful.prototype, "variant", void 0);
    __decorate([
        HostListener('click'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], NglButtonStateful.prototype, "onSelectChange", null);
    __decorate([
        HostListener('focus', ['1']),
        HostListener('blur', ['0']),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], NglButtonStateful.prototype, "onFocusToggle", null);
    NglButtonStateful = __decorate([
        Directive({
            selector: '[nglButtonStateful]',
            providers: [HostService],
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2, HostService])
    ], NglButtonStateful);
    return NglButtonStateful;
}());
export { NglButtonStateful };
