// ng-lightning/esm2015/lib/colorpicker/swatches/colorpicker-swatches.js
import { __decorate, __metadata } from "tslib";
import { Component, ElementRef, Renderer2, ChangeDetectionStrategy, Input, Output, EventEmitter, HostListener, ViewChildren, QueryList, OnChanges } from '@angular/core';
import { LEFT_ARROW, DOWN_ARROW, UP_ARROW, RIGHT_ARROW } from '@angular/cdk/keycodes';
import { trapEvent } from '../../util/util';
import { NglColorpickerSwatchTrigger } from './trigger';
let NglColorpickerSwatches = class NglColorpickerSwatches {
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.hexChange = new EventEmitter();
        this.swatchColors = [];
        this.renderer.addClass(this.el.nativeElement, 'slds-color-picker__swatches');
    }
    ngOnChanges() {
        this.activeIndex = Math.max(this.swatchColors.indexOf(this.hex), 0);
    }
    onSelectViaInteraction(evt) {
        let direction = 0;
        switch (evt.keyCode) {
            case LEFT_ARROW:
            case UP_ARROW:
                direction = -1;
                break;
            case RIGHT_ARROW:
            case DOWN_ARROW:
                direction = 1;
                break;
            default:
                return;
        }
        trapEvent(evt);
        const activeIndex = this.swatchColors.indexOf(this.hex);
        const index = (this.triggers.length + activeIndex + direction) % this.triggers.length;
        const trigger = this.triggers.toArray()[index];
        trigger.focus();
    }
    isSelected(hex) {
        return hex === this.hex;
    }
    onSelect(hex) {
        this.hexChange.emit(hex);
    }
};
NglColorpickerSwatches.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
__decorate([
    Input(),
    __metadata("design:type", String)
], NglColorpickerSwatches.prototype, "hex", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NglColorpickerSwatches.prototype, "hexChange", void 0);
__decorate([
    Input(),
    __metadata("design:type", Array)
], NglColorpickerSwatches.prototype, "swatchColors", void 0);
__decorate([
    ViewChildren(NglColorpickerSwatchTrigger),
    __metadata("design:type", QueryList)
], NglColorpickerSwatches.prototype, "triggers", void 0);
__decorate([
    HostListener('keydown', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [KeyboardEvent]),
    __metadata("design:returntype", void 0)
], NglColorpickerSwatches.prototype, "onSelectViaInteraction", null);
NglColorpickerSwatches = __decorate([
    Component({
        selector: 'ngl-colorpicker-swatches',
        template: "\n<li class=\"slds-color-picker__swatch\" *ngFor=\"let color of swatchColors; let i = index\" role=\"presentation\"><a nglColorpickerSwatchTrigger href=\"javascript:void(0);\" [selected]=\"isSelected(color)\" [attr.tabindex]=\"activeIndex === i ? 0 : -1\" (selectedChange)=\"onSelect(color)\"><span nglColorpickerSwatch [color]=\"color\"></span></a></li>",
        changeDetection: ChangeDetectionStrategy.OnPush,
        styles: [`
    .ngl-color-picker__swatch-selected {
      box-shadow: rgb(117, 112, 112) 1px 1px 1px;
  `]
    }),
    __metadata("design:paramtypes", [ElementRef, Renderer2])
], NglColorpickerSwatches);
export { NglColorpickerSwatches };
