// ng-lightning/esm2015/lib/colorpicker/swatches/trigger.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, Renderer2, Input, Output, EventEmitter, HostListener, HostBinding } from '@angular/core';
let NglColorpickerSwatchTrigger = class NglColorpickerSwatchTrigger {
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.selectedChange = new EventEmitter();
        this.renderer.addClass(this.el.nativeElement, 'slds-color-picker__swatch-trigger');
        this.renderer.setAttribute(this.el.nativeElement, 'role', 'option');
    }
    onSelect() {
        return this.selectedChange.emit();
    }
    focus() {
        this.el.nativeElement.focus();
        this.onSelect();
    }
};
NglColorpickerSwatchTrigger.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
__decorate([
    HostBinding('class.ngl-color-picker__swatch-selected'),
    Input(),
    __metadata("design:type", Boolean)
], NglColorpickerSwatchTrigger.prototype, "selected", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NglColorpickerSwatchTrigger.prototype, "selectedChange", void 0);
__decorate([
    HostListener('click'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], NglColorpickerSwatchTrigger.prototype, "onSelect", null);
NglColorpickerSwatchTrigger = __decorate([
    Directive({
        selector: '[nglColorpickerSwatchTrigger]',
    }),
    __metadata("design:paramtypes", [ElementRef, Renderer2])
], NglColorpickerSwatchTrigger);
export { NglColorpickerSwatchTrigger };
