// ng-lightning/esm5/lib/forms/label.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ChangeDetectionStrategy, TemplateRef, ElementRef, Renderer2, OnInit } from '@angular/core';
import { InputBoolean } from '../util/convert';
var NglFormLabel = /** @class */ (function () {
    function NglFormLabel(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.klass = 'slds-form-element__label';
    }
    NglFormLabel.prototype.ngOnInit = function () {
        this.renderer.addClass(this.element.nativeElement, this.klass);
    };
    NglFormLabel.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input('nglFormLabel'),
        __metadata("design:type", Object)
    ], NglFormLabel.prototype, "label", void 0);
    __decorate([
        Input('nglFormLabelClass'),
        __metadata("design:type", Object)
    ], NglFormLabel.prototype, "klass", void 0);
    __decorate([
        Input(), InputBoolean(),
        __metadata("design:type", Boolean)
    ], NglFormLabel.prototype, "required", void 0);
    NglFormLabel = __decorate([
        Component({
            // tslint:disable-next-line:component-selector
            selector: 'label[nglFormLabel]',
            template: "<abbr class=\"slds-required\" *ngIf=\"required\" title=\"Required\">*</abbr><span [nglInternalOutlet]=\"label\"></span>\n<ng-content></ng-content>",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2])
    ], NglFormLabel);
    return NglFormLabel;
}());
export { NglFormLabel };
