// ng-lightning/esm5/lib/colorpicker/swatches/trigger.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, Renderer2, Input, Output, EventEmitter, HostListener, HostBinding } from '@angular/core';
var NglColorpickerSwatchTrigger = /** @class */ (function () {
    function NglColorpickerSwatchTrigger(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.selectedChange = new EventEmitter();
        this.renderer.addClass(this.el.nativeElement, 'slds-color-picker__swatch-trigger');
        this.renderer.setAttribute(this.el.nativeElement, 'role', 'option');
    }
    NglColorpickerSwatchTrigger.prototype.onSelect = function () {
        return this.selectedChange.emit();
    };
    NglColorpickerSwatchTrigger.prototype.focus = function () {
        this.el.nativeElement.focus();
        this.onSelect();
    };
    NglColorpickerSwatchTrigger.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        HostBinding('class.ngl-color-picker__swatch-selected'),
        Input(),
        __metadata("design:type", Boolean)
    ], NglColorpickerSwatchTrigger.prototype, "selected", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NglColorpickerSwatchTrigger.prototype, "selectedChange", void 0);
    __decorate([
        HostListener('click'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], NglColorpickerSwatchTrigger.prototype, "onSelect", null);
    NglColorpickerSwatchTrigger = __decorate([
        Directive({
            selector: '[nglColorpickerSwatchTrigger]',
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2])
    ], NglColorpickerSwatchTrigger);
    return NglColorpickerSwatchTrigger;
}());
export { NglColorpickerSwatchTrigger };
