// ng-lightning/esm5/lib/radio-group/radio-group.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ChangeDetectionStrategy, HostBinding } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { toBoolean, InputBoolean } from '../util/convert';
import { uniqueId } from '../util/util';
var NglRadioGroup = /** @class */ (function () {
    function NglRadioGroup() {
        this.type = 'list';
        this.uid = uniqueId('radio-group');
        this.type$ = new BehaviorSubject(this.type);
        this.error$ = new BehaviorSubject(this.error);
    }
    Object.defineProperty(NglRadioGroup.prototype, "hasError", {
        get: function () {
            return toBoolean(this.error);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NglRadioGroup.prototype, "errorId", {
        get: function () {
            return "error_" + this.uid;
        },
        enumerable: true,
        configurable: true
    });
    NglRadioGroup.prototype.ngOnChanges = function (changes) {
        if (changes.type) {
            this.type$.next(this.type);
        }
        if (changes.error) {
            this.error$.next(this.hasError ? this.errorId : null);
        }
    };
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglRadioGroup.prototype, "label", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglRadioGroup.prototype, "error", void 0);
    __decorate([
        HostBinding('class.slds-has-error'),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [])
    ], NglRadioGroup.prototype, "hasError", null);
    __decorate([
        Input(), InputBoolean(),
        __metadata("design:type", Boolean)
    ], NglRadioGroup.prototype, "required", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglRadioGroup.prototype, "type", void 0);
    NglRadioGroup = __decorate([
        Component({
            selector: 'ngl-radio-group,[ngl-radio-group]',
            template: "\n<legend class=\"slds-form-element__legend slds-form-element__label\"><abbr class=\"slds-required\" *ngIf=\"required\" title=\"required\">*</abbr><span [nglInternalOutlet]=\"label\"></span></legend>\n<div class=\"slds-form-element__control\">\n  <div class=\"slds-radio_button-group\" *ngIf=\"type === 'button'; else contentTpl\">\n    <ng-container *ngTemplateOutlet=\"contentTpl\"></ng-container>\n  </div>\n</div>\n<div class=\"slds-form-element__help\" *ngIf=\"error\" [id]=\"errorId\">{{error}}</div>\n<ng-template #contentTpl>\n  <ng-content></ng-content>\n</ng-template>",
            changeDetection: ChangeDetectionStrategy.OnPush,
            host: {
                '[class.slds-form-element]': 'true',
            }
        })
    ], NglRadioGroup);
    return NglRadioGroup;
}());
export { NglRadioGroup };
