// ng-lightning/esm5/lib/radio-group/radio-option.js
import { __decorate, __metadata } from "tslib";
import { Component, TemplateRef, Input, ContentChild, ChangeDetectionStrategy, ChangeDetectorRef, HostBinding, OnInit, AfterContentInit, OnDestroy } from '@angular/core';
import { NglRadioGroup } from './radio-group';
import { NglRadioInput } from './input/input';
var NglRadioOption = /** @class */ (function () {
    function NglRadioOption(radioGroup, cd) {
        this.radioGroup = radioGroup;
        this.cd = cd;
        this.subscriptions = [];
    }
    Object.defineProperty(NglRadioOption.prototype, "isTypeList", {
        get: function () {
            return this.type === 'list';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NglRadioOption.prototype, "isTypeButton", {
        get: function () {
            return this.type === 'button';
        },
        enumerable: true,
        configurable: true
    });
    NglRadioOption.prototype.ngOnInit = function () {
        var _this = this;
        this.subscriptions.push(this.radioGroup.type$.subscribe(function (type) {
            _this.type = type;
            _this.cd.detectChanges();
        }), this.radioGroup.error$.subscribe(function (errorId) {
            _this.input.describedBy = errorId;
        }));
    };
    NglRadioOption.prototype.ngAfterContentInit = function () {
        this.input.name = this.radioGroup.uid;
    };
    NglRadioOption.prototype.ngOnDestroy = function () {
        this.subscriptions.forEach(function (s) { return s.unsubscribe(); });
    };
    NglRadioOption.ctorParameters = function () { return [
        { type: NglRadioGroup },
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglRadioOption.prototype, "label", void 0);
    __decorate([
        ContentChild(NglRadioInput, { static: true }),
        __metadata("design:type", NglRadioInput)
    ], NglRadioOption.prototype, "input", void 0);
    __decorate([
        HostBinding('class.slds-radio'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], NglRadioOption.prototype, "isTypeList", null);
    __decorate([
        HostBinding('class.slds-button'),
        HostBinding('class.slds-radio_button'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], NglRadioOption.prototype, "isTypeButton", null);
    NglRadioOption = __decorate([
        Component({
            selector: 'ngl-radio-option',
            template: "\n<ng-content></ng-content>\n<label class=\"slds-radio__label\" *ngIf=\"type === 'list'\" [attr.for]=\"input.id\"><span class=\"slds-radio_faux\"></span><span class=\"slds-form-element__label\" [nglInternalOutlet]=\"label\"></span></label>\n<label class=\"slds-radio_button__label\" *ngIf=\"type === 'button'\" [attr.for]=\"input.id\"><span class=\"slds-radio_faux\" [nglInternalOutlet]=\"label\"></span></label>",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        __metadata("design:paramtypes", [NglRadioGroup, ChangeDetectorRef])
    ], NglRadioOption);
    return NglRadioOption;
}());
export { NglRadioOption };
