// ng-lightning/esm2015/lib/input/element/element.js
import { __decorate, __metadata } from "tslib";
import { Directive, HostBinding, ElementRef, Input, Renderer2 } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { toBoolean } from '../../util/convert';
import { uniqueId } from '../../util/util';
let NglInputElement = class NglInputElement {
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.ɵRequiredSubject = new BehaviorSubject(false);
        if (!this.el.nativeElement.id) {
            this.renderer.setAttribute(this.el.nativeElement, 'id', uniqueId('input'));
        }
    }
    set required(required) {
        this.ɵRequiredSubject.next(toBoolean(required));
    }
    get id() {
        return this.el.nativeElement.id;
    }
};
NglInputElement.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
__decorate([
    HostBinding('attr.aria-describedby'),
    __metadata("design:type", String)
], NglInputElement.prototype, "describedBy", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], NglInputElement.prototype, "required", null);
NglInputElement = __decorate([
    Directive({
        selector: 'input[ngl]:not([type=checkbox]):not([type=radio])',
        host: {
            '[class.slds-input]': 'true',
        }
    }),
    __metadata("design:paramtypes", [ElementRef, Renderer2])
], NglInputElement);
export { NglInputElement };
