// ng-lightning/esm2015/lib/tabs/tab-verbose.js
var NglTabVerbose_1;
import { __decorate, __metadata } from "tslib";
import { Directive, TemplateRef, ContentChild, AfterContentInit } from '@angular/core';
import { NglTab } from './tab';
/*
 * <ngl-tab [label="..."]>
 *    <ng-template ngl-tab-label>...</ng-template>
 *    <ng-template ngl-tab-content>
 *       Content goes here...
 *    </ng-template>
 * </ngl-tab>
 */
// tslint:disable-next-line:directive-selector
let NglTabLabel = class NglTabLabel {
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
};
NglTabLabel.ctorParameters = () => [
    { type: TemplateRef }
];
NglTabLabel = __decorate([
    Directive({ selector: '[ngl-tab-label]' }),
    __metadata("design:paramtypes", [TemplateRef])
], NglTabLabel);
export { NglTabLabel };
// tslint:disable-next-line:directive-selector
let NglTabContent = class NglTabContent {
    constructor(templateRef) {
        this.templateRef = templateRef;
    }
};
NglTabContent.ctorParameters = () => [
    { type: TemplateRef }
];
NglTabContent = __decorate([
    Directive({ selector: '[ngl-tab-content]' }),
    __metadata("design:paramtypes", [TemplateRef])
], NglTabContent);
export { NglTabContent };
let NglTabVerbose = NglTabVerbose_1 = class NglTabVerbose extends NglTab {
    ngAfterContentInit() {
        if (this.labelTemplate) {
            this.label = this.labelTemplate.templateRef;
        }
        this.templateRef = this.contentTemplate.templateRef;
    }
};
__decorate([
    ContentChild(NglTabContent),
    __metadata("design:type", NglTabContent)
], NglTabVerbose.prototype, "contentTemplate", void 0);
__decorate([
    ContentChild(NglTabLabel),
    __metadata("design:type", NglTabLabel)
], NglTabVerbose.prototype, "labelTemplate", void 0);
NglTabVerbose = NglTabVerbose_1 = __decorate([
    Directive({
        // tslint:disable-next-line:directive-selector
        selector: 'ngl-tab',
        providers: [{ provide: NglTab, useExisting: NglTabVerbose_1 }],
    })
], NglTabVerbose);
export { NglTabVerbose };
