// ng-lightning/esm2015/lib/tabs/tab.js
import { __decorate, __metadata, __param } from "tslib";
import { Directive, Input, TemplateRef, Output, EventEmitter, Optional } from '@angular/core';
import { uniqueId } from '../util/util';
/*
 * <ng-template ngl-tab label="...">
 *    Content goes here...
 * </ng-template>
 */
let NglTab = class NglTab {
    constructor(templateRef) {
        this.templateRef = templateRef;
        this.activate = new EventEmitter();
        this.deactivate = new EventEmitter();
        this.uid = uniqueId('tab');
        this._active = false;
    }
    set active(active) {
        if (active === this._active) {
            return;
        }
        this._active = active;
        if (active) {
            this.activate.emit(this);
        }
        else {
            this.deactivate.emit(this);
        }
    }
    get active() {
        return this._active;
    }
};
NglTab.ctorParameters = () => [
    { type: TemplateRef, decorators: [{ type: Optional }] }
];
__decorate([
    Input(),
    __metadata("design:type", String)
], NglTab.prototype, "id", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], NglTab.prototype, "label", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NglTab.prototype, "activate", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NglTab.prototype, "deactivate", void 0);
NglTab = __decorate([
    Directive({
        // tslint:disable-next-line:directive-selector
        selector: '[ngl-tab]',
        exportAs: 'nglTab',
    }),
    __param(0, Optional()),
    __metadata("design:paramtypes", [TemplateRef])
], NglTab);
export { NglTab };
