// ng-lightning/esm5/lib/checkboxes/toggle/checkbox-toggle.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ChangeDetectionStrategy, ContentChild, ChangeDetectorRef, AfterContentInit, TemplateRef, HostBinding, OnDestroy } from '@angular/core';
import { NglCheckboxInput } from '../input/input';
import { toBoolean } from '../../util/convert';
var NglCheckboxToggle = /** @class */ (function () {
    function NglCheckboxToggle(cd) {
        this.cd = cd;
        this.enabledText = 'Enabled';
        this.disabledText = 'Disabled';
    }
    Object.defineProperty(NglCheckboxToggle.prototype, "hasError", {
        get: function () {
            return toBoolean(this.error);
        },
        enumerable: true,
        configurable: true
    });
    NglCheckboxToggle.prototype.ngAfterContentInit = function () {
        var _this = this;
        if (!this.input) {
            throw Error("[ng-lightning] Couldn't find an <input type=\"checkbox\"> with [ngl] attribute inside NglCheckboxToggle");
        }
        this.ɵRequiredSubscription = this.input.ɵRequiredSubject.subscribe(function (required) {
            _this.required = required;
            _this.cd.detectChanges();
        });
        this.uid = this.input.id + "_toggle";
        this.input.describedBy = this.uid;
        this.cd.detectChanges();
    };
    NglCheckboxToggle.prototype.ngOnDestroy = function () {
        if (this.ɵRequiredSubscription) {
            this.ɵRequiredSubscription.unsubscribe();
            this.ɵRequiredSubscription = null;
        }
    };
    NglCheckboxToggle.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        ContentChild(NglCheckboxInput, { static: true }),
        __metadata("design:type", NglCheckboxInput)
    ], NglCheckboxToggle.prototype, "input", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglCheckboxToggle.prototype, "label", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglCheckboxToggle.prototype, "error", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglCheckboxToggle.prototype, "enabledText", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglCheckboxToggle.prototype, "disabledText", void 0);
    __decorate([
        HostBinding('class.slds-has-error'),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [])
    ], NglCheckboxToggle.prototype, "hasError", null);
    NglCheckboxToggle = __decorate([
        Component({
            selector: 'ngl-checkbox-toggle',
            template: "\n<label class=\"slds-checkbox_toggle slds-grid\"><abbr class=\"slds-required\" *ngIf=\"required\" title=\"required\">*</abbr><span class=\"slds-form-element__label slds-m-bottom_none\" [nglInternalOutlet]=\"label\"></span>\n  <ng-content></ng-content><span class=\"slds-checkbox_faux_container\" [id]=\"uid\" aria-live=\"assertive\"><span class=\"slds-checkbox_faux\"></span><span class=\"slds-checkbox_on\">{{enabledText}}</span><span class=\"slds-checkbox_off\">{{disabledText}}</span></span>\n</label>\n<div class=\"slds-form-element__help\" *ngIf=\"error\">{{error}}</div>",
            changeDetection: ChangeDetectionStrategy.OnPush,
            host: {
                '[class.slds-form-element]': 'true',
            }
        }),
        __metadata("design:paramtypes", [ChangeDetectorRef])
    ], NglCheckboxToggle);
    return NglCheckboxToggle;
}());
export { NglCheckboxToggle };
