// ng-lightning/esm5/lib/avatar/avatar.js
import { __decorate, __metadata, __read } from "tslib";
import { Component, Input, Output, EventEmitter, ChangeDetectionStrategy, ElementRef, Renderer2, HostBinding, OnInit, OnChanges } from '@angular/core';
import { HostService } from '../common/host/host.service';
var NglAvatar = /** @class */ (function () {
    function NglAvatar(element, renderer, hostService) {
        this.element = element;
        this.hostService = hostService;
        this.src = '';
        this.alternativeText = '';
        this.fallbackIconName = 'standard:user';
        this.error = new EventEmitter();
        this._imgError = false;
        renderer.addClass(element.nativeElement, 'slds-avatar');
    }
    NglAvatar.prototype.fallbackIconClass = function () {
        var _a = __read(this.fallbackIconName.split(':'), 2), category = _a[0], icon = _a[1];
        return "slds-icon-" + category + "-" + icon;
    };
    Object.defineProperty(NglAvatar.prototype, "shouldShowImage", {
        get: function () {
            return this.src && !this._imgError;
        },
        enumerable: true,
        configurable: true
    });
    NglAvatar.prototype.onImgError = function () {
        this._imgError = true;
        this.error.emit();
    };
    NglAvatar.prototype.ngOnInit = function () {
        this.setHostClass();
    };
    NglAvatar.prototype.ngOnChanges = function () {
        this.setHostClass();
    };
    NglAvatar.prototype.setHostClass = function () {
        var _a;
        this.hostService.updateClass(this.element, (_a = {},
            _a["slds-avatar_" + (this.size || 'medium')] = true,
            _a["slds-avatar_" + (this.variant || 'rectangle')] = true,
            _a));
    };
    NglAvatar.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: HostService }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglAvatar.prototype, "src", void 0);
    __decorate([
        HostBinding('attr.title'),
        Input(),
        __metadata("design:type", Object)
    ], NglAvatar.prototype, "alternativeText", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglAvatar.prototype, "size", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglAvatar.prototype, "variant", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglAvatar.prototype, "initials", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglAvatar.prototype, "fallbackIconName", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NglAvatar.prototype, "error", void 0);
    NglAvatar = __decorate([
        Component({
            selector: 'ngl-avatar',
            changeDetection: ChangeDetectionStrategy.OnPush,
            template: "<img *ngIf=\"shouldShowImage; else template_initials\" [src]=\"src\" [alt]=\"alternativeText\" (error)=\"onImgError()\">\n<ng-template #template_initials><abbr class=\"slds-avatar__initials\" [ngClass]=\"fallbackIconClass()\">{{ initials }}</abbr></ng-template>",
            providers: [HostService]
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2, HostService])
    ], NglAvatar);
    return NglAvatar;
}());
export { NglAvatar };
