// ng-lightning/esm5/lib/colorpicker/custom/colorpicker-custom.js
import { __decorate, __metadata } from "tslib";
import { Component, ElementRef, Renderer2, ChangeDetectionStrategy, Input, Output, EventEmitter, OnChanges, SimpleChanges } from '@angular/core';
import { getHexFromHsv, getHsvFromHex } from '../util';
var NglColorpickerCustom = /** @class */ (function () {
    function NglColorpickerCustom(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.hsvChange = new EventEmitter();
        this.renderer.addClass(this.el.nativeElement, 'slds-color-picker__custom');
    }
    NglColorpickerCustom.prototype.ngOnChanges = function (changes) {
        if (changes.hsv) {
            this.hex = getHexFromHsv(this.hsv);
        }
    };
    NglColorpickerCustom.prototype.onHsvChange = function ($event) {
        this.hsvChange.emit($event);
    };
    NglColorpickerCustom.prototype.onHexChange = function (hex) {
        var hsv = getHsvFromHex(hex);
        this.hsvChange.emit(hsv);
    };
    NglColorpickerCustom.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglColorpickerCustom.prototype, "hsv", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NglColorpickerCustom.prototype, "hsvChange", void 0);
    NglColorpickerCustom = __decorate([
        Component({
            selector: 'ngl-colorpicker-custom',
            template: "\n<ngl-colorpicker-range [hsv]=\"hsv\" (hsvChange)=\"onHsvChange($event)\"></ngl-colorpicker-range>\n<ngl-colorpicker-inputs [hex]=\"hex\" (hexChange)=\"onHexChange($event)\"></ngl-colorpicker-inputs>",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2])
    ], NglColorpickerCustom);
    return NglColorpickerCustom;
}());
export { NglColorpickerCustom };
