// ng-lightning/esm5/lib/tabs/tab.js
import { __decorate, __metadata, __param } from "tslib";
import { Directive, Input, TemplateRef, Output, EventEmitter, Optional } from '@angular/core';
import { uniqueId } from '../util/util';
/*
 * <ng-template ngl-tab label="...">
 *    Content goes here...
 * </ng-template>
 */
var NglTab = /** @class */ (function () {
    function NglTab(templateRef) {
        this.templateRef = templateRef;
        this.activate = new EventEmitter();
        this.deactivate = new EventEmitter();
        this.uid = uniqueId('tab');
        this._active = false;
    }
    Object.defineProperty(NglTab.prototype, "active", {
        get: function () {
            return this._active;
        },
        set: function (active) {
            if (active === this._active) {
                return;
            }
            this._active = active;
            if (active) {
                this.activate.emit(this);
            }
            else {
                this.deactivate.emit(this);
            }
        },
        enumerable: true,
        configurable: true
    });
    NglTab.ctorParameters = function () { return [
        { type: TemplateRef, decorators: [{ type: Optional }] }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglTab.prototype, "id", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglTab.prototype, "label", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NglTab.prototype, "activate", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NglTab.prototype, "deactivate", void 0);
    NglTab = __decorate([
        Directive({
            // tslint:disable-next-line:directive-selector
            selector: '[ngl-tab]',
            exportAs: 'nglTab',
        }),
        __param(0, Optional()),
        __metadata("design:paramtypes", [TemplateRef])
    ], NglTab);
    return NglTab;
}());
export { NglTab };
