// ng-lightning/esm5/lib/tabs/tabs.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, QueryList, ContentChildren, Output, EventEmitter, ElementRef, Renderer2, AfterContentInit } from '@angular/core';
import { isInt } from '../util/util';
import { NglTab } from './tab';
import { InputBoolean } from '../util/convert';
var NglTabs = /** @class */ (function () {
    function NglTabs(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.selectedChange = new EventEmitter();
        /**
         * Whether every tab's content is instantiated when visible, and destroyed when hidden.
         */
        this.lazy = true;
        this.renderer.addClass(this.element.nativeElement, "slds-tabs_" + this.variant);
    }
    Object.defineProperty(NglTabs.prototype, "variant", {
        get: function () {
            return this._variant || 'default';
        },
        set: function (variant) {
            var el = this.element.nativeElement;
            this.renderer.removeClass(el, "slds-tabs_" + this.variant);
            this._variant = variant;
            this.renderer.addClass(el, "slds-tabs_" + this.variant);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NglTabs.prototype, "setSelected", {
        set: function (selected) {
            if (selected === this.selected) {
                return;
            }
            this.selected = selected;
            if (!this.tabs) {
                return;
            } // Wait for content to initialize
            this.activate();
        },
        enumerable: true,
        configurable: true
    });
    NglTabs.prototype.ngAfterContentInit = function () {
        var _this = this;
        // Initial selection after all tabs are created
        this.activate();
        if (!this.activeTab) {
            setTimeout(function () { return _this.select(_this.tabs.first); });
        }
    };
    NglTabs.prototype.select = function (tab) {
        this.selectedChange.emit(tab);
    };
    NglTabs.prototype.move = function (evt, moves) {
        evt.preventDefault();
        var tabs = this.tabs.toArray();
        var selectedIndex = tabs.indexOf(this.activeTab);
        this.select(tabs[(tabs.length + selectedIndex + moves) % tabs.length]);
    };
    NglTabs.prototype.tabClass = function (tab) {
        var _a;
        return _a = {},
            _a["slds-tabs_" + this.variant + "__content"] = true,
            _a["slds-" + (tab.active ? 'show' : 'hide')] = true,
            _a;
    };
    NglTabs.prototype.trackByTab = function (index, tab) {
        return tab.uid;
    };
    NglTabs.prototype.activate = function () {
        if (this.activeTab) {
            this.activeTab.active = false;
        }
        this.activeTab = this.findTab();
        if (this.activeTab) {
            this.activeTab.active = true;
        }
    };
    NglTabs.prototype.findTab = function (value) {
        if (value === void 0) { value = this.selected; }
        if (value instanceof NglTab) {
            return value;
        }
        if (isInt(value)) {
            return this.tabs.toArray()[+value];
        }
        return this.tabs.toArray().find(function (t) {
            return t.id && t.id === value;
        });
    };
    NglTabs.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], NglTabs.prototype, "variant", null);
    __decorate([
        ContentChildren(NglTab),
        __metadata("design:type", QueryList)
    ], NglTabs.prototype, "tabs", void 0);
    __decorate([
        Input('selected'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], NglTabs.prototype, "setSelected", null);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NglTabs.prototype, "selectedChange", void 0);
    __decorate([
        Input(), InputBoolean(),
        __metadata("design:type", Object)
    ], NglTabs.prototype, "lazy", void 0);
    NglTabs = __decorate([
        Component({
            selector: 'ngl-tabset',
            template: "\n<ul [ngClass]=\"'slds-tabs_' + variant + '__nav'\" role=\"tablist\" (keydown.ArrowLeft)=\"move($event, -1)\" (keydown.ArrowRight)=\"move($event, 1)\">\n  <li *ngFor=\"let tab of tabs; trackBy: trackByTab\" [ngClass]=\"'slds-tabs_' + variant + '__item'\" [class.slds-is-active]=\"tab.active\" [id]=\"tab.uid + '__item'\" [attr.aria-controls]=\"tab.uid\" (click)=\"select(tab)\" role=\"presentation\"><a [nglInternalOutlet]=\"tab.label\" [ngClass]=\"'slds-tabs_' + variant + '__link'\" role=\"tab\" [attr.aria-selected]=\"tab.active\" [attr.tabindex]=\"tab.active ? 0 : -1\"></a></li>\n</ul>\n<div *ngFor=\"let tab of tabs; trackBy: trackByTab\" [id]=\"tab.uid\" [attr.aria-labelledby]=\"tab.uid + '__item'\" [ngClass]=\"tabClass(tab)\" role=\"tabpanel\">\n  <ng-container *ngIf=\"!lazy || tab.active\">\n    <ng-template [ngTemplateOutlet]=\"tab?.templateRef\"></ng-template>\n  </ng-container>\n</div>"
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2])
    ], NglTabs);
    return NglTabs;
}());
export { NglTabs };
