// ng-lightning/esm5/lib/carousel/carousel-image.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ChangeDetectionStrategy, TemplateRef, Renderer2, ElementRef } from '@angular/core';
import { uniqueId } from '../util/util';
var NglCarouselImage = /** @class */ (function () {
    function NglCarouselImage(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.uid = uniqueId('carousel-image');
        this.renderer.setAttribute(this.el.nativeElement, 'id', this.uid);
        this.renderer.addClass(this.el.nativeElement, 'slds-carousel__panel');
        this.renderer.setAttribute(this.el.nativeElement, 'role', 'tabpanel');
    }
    Object.defineProperty(NglCarouselImage.prototype, "labelledby", {
        set: function (labelledby) {
            this.renderer.setAttribute(this.el.nativeElement, 'aria-labelledby', labelledby);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NglCarouselImage.prototype, "active", {
        set: function (active) {
            this.renderer.setAttribute(this.el.nativeElement, 'aria-hidden', "" + !active);
        },
        enumerable: true,
        configurable: true
    });
    NglCarouselImage.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglCarouselImage.prototype, "src", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglCarouselImage.prototype, "header", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglCarouselImage.prototype, "description", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglCarouselImage.prototype, "alternativeText", void 0);
    NglCarouselImage = __decorate([
        Component({
            selector: 'ngl-carousel-image',
            template: "<a class=\"slds-carousel__panel-action slds-text-link_reset\" href=\"javascript:void(0);\" [attr.tabindex]=\"active ? 0 : -1\">\n  <div class=\"slds-carousel__image\"><img [src]=\"src\" [attr.alt]=\"alternativeText || null\"></div>\n  <div class=\"slds-carousel__content\">\n    <h2 class=\"slds-carousel__content-title\" [nglInternalOutlet]=\"header\"></h2>\n    <p [nglInternalOutlet]=\"description\"></p>\n  </div></a>",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2])
    ], NglCarouselImage);
    return NglCarouselImage;
}());
export { NglCarouselImage };
