// ng-lightning/esm5/lib/carousel/carousel-indicator.js
import { __decorate, __metadata } from "tslib";
import { Directive, HostBinding, Input, ElementRef, OnChanges, SimpleChanges } from '@angular/core';
import { InputBoolean } from '../util/convert';
import { NglCarouselImage } from './carousel-image';
import { uniqueId } from '../util/util';
var NglCarouselIndicator = /** @class */ (function () {
    function NglCarouselIndicator(el) {
        this.el = el;
        this.uid = uniqueId('carousel-indicator');
    }
    Object.defineProperty(NglCarouselIndicator.prototype, "tabindex", {
        get: function () {
            return this.isActive ? 0 : -1;
        },
        enumerable: true,
        configurable: true
    });
    NglCarouselIndicator.prototype.ngOnChanges = function (changes) {
        this.image.active = this.isActive;
        if (changes.image) {
            this.image.labelledby = this.uid;
        }
    };
    NglCarouselIndicator.prototype.focus = function () {
        this.el.nativeElement.focus();
    };
    NglCarouselIndicator.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        HostBinding('class.slds-is-active'),
        HostBinding('attr.aria-selected'),
        Input(), InputBoolean(),
        __metadata("design:type", Object)
    ], NglCarouselIndicator.prototype, "isActive", void 0);
    __decorate([
        HostBinding('attr.tabindex'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], NglCarouselIndicator.prototype, "tabindex", null);
    __decorate([
        Input(),
        __metadata("design:type", NglCarouselImage)
    ], NglCarouselIndicator.prototype, "image", void 0);
    __decorate([
        HostBinding('attr.id'),
        __metadata("design:type", Object)
    ], NglCarouselIndicator.prototype, "uid", void 0);
    NglCarouselIndicator = __decorate([
        Directive({
            selector: '[nglCarouselIndicator]',
        }),
        __metadata("design:paramtypes", [ElementRef])
    ], NglCarouselIndicator);
    return NglCarouselIndicator;
}());
export { NglCarouselIndicator };
