// ng-lightning/esm2015/lib/buttons/button-stateful.js
import { __decorate, __metadata } from "tslib";
import { Directive, Input, Output, EventEmitter, HostListener, ElementRef, Renderer2, OnInit, OnChanges } from '@angular/core';
import { InputBoolean } from '../util/convert';
import { HostService } from '../common/host/host.service';
let NglButtonStateful = class NglButtonStateful {
    constructor(el, renderer, hostService) {
        this.el = el;
        this.renderer = renderer;
        this.hostService = hostService;
        /**
         * Triggered when the button is clicked.
         */
        this.stateChange = new EventEmitter();
        /**
         * Appearance.
         */
        this.variant = 'neutral';
        this.renderer.addClass(this.el.nativeElement, 'slds-button');
        this.renderer.addClass(this.el.nativeElement, 'slds-button_stateful');
        this.renderer.setAttribute(this.el.nativeElement, 'aria-live', 'assertive');
    }
    onSelectChange() {
        this.stateChange.emit(!this.state);
    }
    onFocusToggle(focused) {
        this.focused = !!+focused;
        if (!this.focused) {
            this.setHostClass();
        }
    }
    ngOnInit() {
        this.setHostClass();
    }
    ngOnChanges() {
        this.setHostClass();
    }
    setHostClass() {
        this.hostService.updateClass(this.el, {
            [`slds-button_${this.variant === 'text' ? 'reset' : this.variant}`]: !!this.variant,
            [`slds-is-selected-clicked`]: this.state && this.focused,
            [`slds-is-selected`]: this.state && !this.focused,
            [`slds-not-selected`]: !this.state,
        });
    }
};
NglButtonStateful.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: HostService }
];
__decorate([
    Input(), InputBoolean(),
    __metadata("design:type", Boolean)
], NglButtonStateful.prototype, "state", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NglButtonStateful.prototype, "stateChange", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], NglButtonStateful.prototype, "variant", void 0);
__decorate([
    HostListener('click'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], NglButtonStateful.prototype, "onSelectChange", null);
__decorate([
    HostListener('focus', ['1']),
    HostListener('blur', ['0']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", void 0)
], NglButtonStateful.prototype, "onFocusToggle", null);
NglButtonStateful = __decorate([
    Directive({
        selector: '[nglButtonStateful]',
        providers: [HostService],
    }),
    __metadata("design:paramtypes", [ElementRef, Renderer2, HostService])
], NglButtonStateful);
export { NglButtonStateful };
