// ng-lightning/esm5/lib/tooltips/tooltip.js
import { __decorate, __metadata } from "tslib";
import { Component, ChangeDetectionStrategy, ChangeDetectorRef, TemplateRef, ElementRef, Renderer2 } from '@angular/core';
import { POSITION_MAP, getPlacementStyles } from '../util/overlay-position';
import { HostService } from '../common/host/host.service';
import { OnChange } from '../util/property-watch-decorator';
import { ngClassCombine } from '../util/util';
var NglTooltip = /** @class */ (function () {
    function NglTooltip(element, renderer, hostService, cd) {
        this.element = element;
        this.renderer = renderer;
        this.hostService = hostService;
        this.cd = cd;
        this.renderer.addClass(this.element.nativeElement, 'slds-popover');
        this.renderer.addClass(this.element.nativeElement, 'slds-popover_tooltip');
        this.renderer.setAttribute(this.element.nativeElement, 'role', 'tooltip');
    }
    NglTooltip.prototype.nglOnPropertyChange = function (prop) {
        if (prop === 'uid') {
            this.renderer.setAttribute(this.element.nativeElement, 'id', this.uid);
        }
        else if (prop === 'placement') {
            this.nubbin = POSITION_MAP[this.placement].nubbin;
            this.setHostClass();
        }
        else if (prop === 'template') {
            this.cd.markForCheck();
        }
        else if (prop === 'tooltipClass') {
            this.setHostClass();
        }
    };
    NglTooltip.prototype.setHostClass = function () {
        var _a;
        this.hostService.updateClass(this.element, ngClassCombine(this.tooltipClass, (_a = {},
            _a["slds-nubbin_" + this.nubbin] = true,
            _a)));
        this.hostService.updateStyle(this.element, getPlacementStyles(this.nubbin));
    };
    NglTooltip.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: HostService },
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        OnChange(),
        __metadata("design:type", Object)
    ], NglTooltip.prototype, "template", void 0);
    __decorate([
        OnChange(),
        __metadata("design:type", String)
    ], NglTooltip.prototype, "placement", void 0);
    __decorate([
        OnChange(),
        __metadata("design:type", String)
    ], NglTooltip.prototype, "uid", void 0);
    __decorate([
        OnChange(),
        __metadata("design:type", Object)
    ], NglTooltip.prototype, "tooltipClass", void 0);
    NglTooltip = __decorate([
        Component({
            // tslint:disable-next-line:component-selector
            selector: 'div[ngl-tooltip]',
            template: "\n<div class=\"slds-popover__body\" [nglInternalOutlet]=\"template\"></div>",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [HostService]
        }),
        __metadata("design:paramtypes", [ElementRef,
            Renderer2,
            HostService,
            ChangeDetectorRef])
    ], NglTooltip);
    return NglTooltip;
}());
export { NglTooltip };
