// ng-lightning/esm5/lib/checkboxes/button/checkbox-button.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ChangeDetectionStrategy, ContentChild, ChangeDetectorRef, AfterContentInit, TemplateRef } from '@angular/core';
import { NglCheckboxInput } from '../input/input';
var NglCheckboxButton = /** @class */ (function () {
    function NglCheckboxButton(cd) {
        this.cd = cd;
    }
    NglCheckboxButton.prototype.ngAfterContentInit = function () {
        if (!this.input) {
            throw Error("[ng-lightning] Couldn't find an <input type=\"checkbox\"> with [ngl] attribute inside " + this);
        }
        this._uid = this.input.id;
        this.cd.detectChanges();
        this.input.addClass('slds-assistive-text');
    };
    NglCheckboxButton.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        ContentChild(NglCheckboxInput, { static: true }),
        __metadata("design:type", NglCheckboxInput)
    ], NglCheckboxButton.prototype, "input", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglCheckboxButton.prototype, "label", void 0);
    NglCheckboxButton = __decorate([
        Component({
            selector: 'ngl-checkbox-button',
            template: "\n<ng-content></ng-content>\n<label class=\"slds-checkbox_faux\" [attr.for]=\"_uid\"><span class=\"slds-assistive-text\" [nglInternalOutlet]=\"label\"></span></label>",
            changeDetection: ChangeDetectionStrategy.OnPush,
            host: {
                '[class.slds-checkbox_add-button]': 'true',
            }
        }),
        __metadata("design:paramtypes", [ChangeDetectorRef])
    ], NglCheckboxButton);
    return NglCheckboxButton;
}());
export { NglCheckboxButton };
