// ng-lightning/esm5/lib/common/overlay/overlay-outside.js
import { __decorate, __metadata, __param } from "tslib";
import { Directive, Output, EventEmitter, OnInit, OnDestroy, NgZone, Self } from '@angular/core';
import { CdkConnectedOverlay, ScrollDispatcher } from '@angular/cdk/overlay';
var NglOverlaynglOverlayScrolledOutsideViewDirective = /** @class */ (function () {
    function NglOverlaynglOverlayScrolledOutsideViewDirective(cdkOverlay, ngZone, scrollDispatcher) {
        this.cdkOverlay = cdkOverlay;
        this.ngZone = ngZone;
        this.scrollDispatcher = scrollDispatcher;
        this.overlayOutside = new EventEmitter();
    }
    NglOverlaynglOverlayScrolledOutsideViewDirective.prototype.ngOnInit = function () {
        var _this = this;
        var elementRef = this.cdkOverlay.origin.elementRef;
        var scrollableAncestors = this.scrollDispatcher.getAncestorScrollContainers(elementRef).map(function (container) { return container.getElementRef(); });
        if (!scrollableAncestors || !scrollableAncestors.length)
            return;
        this.subscription = this.cdkOverlay.positionChange.subscribe(function () {
            var bounds = elementRef.nativeElement.getBoundingClientRect();
            for (var i = 0, n = scrollableAncestors.length; i < n; i++) {
                var ancestorsBounds = scrollableAncestors[i].nativeElement.getBoundingClientRect();
                if (isElementOutside(bounds, ancestorsBounds)) {
                    _this.ngZone.run(function () { return _this.overlayOutside.emit(); });
                    return;
                }
            }
        });
    };
    NglOverlaynglOverlayScrolledOutsideViewDirective.prototype.ngOnDestroy = function () {
        if (this.subscription) {
            this.subscription.unsubscribe();
            this.subscription = null;
        }
    };
    NglOverlaynglOverlayScrolledOutsideViewDirective.ctorParameters = function () { return [
        { type: CdkConnectedOverlay, decorators: [{ type: Self }] },
        { type: NgZone },
        { type: ScrollDispatcher }
    ]; };
    __decorate([
        Output('nglOverlayScrolledOutsideView'),
        __metadata("design:type", EventEmitter)
    ], NglOverlaynglOverlayScrolledOutsideViewDirective.prototype, "overlayOutside", void 0);
    NglOverlaynglOverlayScrolledOutsideViewDirective = __decorate([
        Directive({
            selector: '[nglOverlayScrolledOutsideView]'
        }),
        __param(0, Self()),
        __metadata("design:paramtypes", [CdkConnectedOverlay,
            NgZone,
            ScrollDispatcher])
    ], NglOverlaynglOverlayScrolledOutsideViewDirective);
    return NglOverlaynglOverlayScrolledOutsideViewDirective;
}());
export { NglOverlaynglOverlayScrolledOutsideViewDirective };
/**
 * Gets whether an element is scrolled outside of view by its parent scrolling container.
 * @param element Dimensions of the element (from getBoundingClientRect)
 * @param container Dimensions of element's scrolling container (from getBoundingClientRect)
 * @returns Whether the element is scrolled out of view
 */
export function isElementOutside(element, container) {
    return (element.bottom < container.top || element.top > container.bottom ||
        element.right < container.left || element.left > container.right);
}
