// ng-lightning/esm2015/lib/textarea/input/input.js
import { __decorate, __metadata } from "tslib";
import { Directive, HostBinding, ElementRef, Input, Renderer2 } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { toBoolean } from '../../util/convert';
import { uniqueId } from '../../util/util';
let NglTextareaInput = class NglTextareaInput {
    constructor(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.ɵRequiredSubject = new BehaviorSubject(false);
        if (!this.el.nativeElement.id) {
            this.renderer.setAttribute(this.el.nativeElement, 'id', uniqueId('textarea'));
        }
    }
    set required(required) {
        this.ɵRequiredSubject.next(toBoolean(required));
    }
    get id() {
        return this.el.nativeElement.id;
    }
};
NglTextareaInput.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
__decorate([
    HostBinding('attr.aria-describedby'),
    __metadata("design:type", String)
], NglTextareaInput.prototype, "describedBy", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], NglTextareaInput.prototype, "required", null);
NglTextareaInput = __decorate([
    Directive({
        selector: 'textarea[ngl]',
        host: {
            '[class.slds-textarea]': 'true',
        },
    }),
    __metadata("design:paramtypes", [ElementRef, Renderer2])
], NglTextareaInput);
export { NglTextareaInput };
