// ng-lightning/esm5/lib/comboboxes/combobox-option.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ChangeDetectionStrategy, OnDestroy, ElementRef, Renderer2, HostListener, ChangeDetectorRef, NgZone } from '@angular/core';
import { uniqueId, trapEvent, menuItemScroll } from '../util/util';
import { InputBoolean } from '../util/convert';
import { NglComboboxService } from './combobox.service';
var NglComboboxOption = /** @class */ (function () {
    function NglComboboxOption(element, service, cd, ngZone, renderer) {
        this.element = element;
        this.service = service;
        this.cd = cd;
        this.ngZone = ngZone;
        this.disabled = false;
        this.uid = uniqueId('combo-option');
        this._active = false;
        // Flag to disable scrolling into view when option is activated using mouse
        this.disableNextScrollIntoView = false;
        this.destroyed = false;
        renderer.addClass(element.nativeElement, 'slds-listbox__item');
        renderer.setAttribute(element.nativeElement, 'role', 'presentation');
    }
    Object.defineProperty(NglComboboxOption.prototype, "active", {
        get: function () {
            return this._active;
        },
        // Whether or not the option is currently active and ready to be selected
        set: function (active) {
            if (this.active === active || this.destroyed) {
                return;
            }
            this._active = active;
            this.cd.detectChanges();
            if (active) {
                this.service.combobox.inputEl.setAriaActiveDescendant(this.uid);
                this.scrollIntoView();
            }
            else {
                clearTimeout(this.scrollTimer);
            }
        },
        enumerable: true,
        configurable: true
    });
    NglComboboxOption.prototype.onSelectViaInteraction = function (evt) {
        trapEvent(evt);
        if (!this.disabled) {
            this.service.combobox.onOptionSelection(this);
        }
    };
    NglComboboxOption.prototype.hover = function () {
        if (!this.disabled) {
            this.disableNextScrollIntoView = true;
            this.service.combobox.keyManager.setActiveItem(this);
        }
    };
    NglComboboxOption.prototype.setActiveStyles = function () {
        this.active = true;
    };
    NglComboboxOption.prototype.setInactiveStyles = function () {
        this.active = false;
    };
    NglComboboxOption.prototype.scrollIntoView = function () {
        var _this = this;
        if (this.disableNextScrollIntoView) {
            this.disableNextScrollIntoView = false;
            return;
        }
        this.ngZone.runOutsideAngular(function () {
            _this.scrollTimer = setTimeout(function () {
                var li = _this.element.nativeElement;
                menuItemScroll(li.parentElement.parentElement, li);
            }, 0);
        });
    };
    NglComboboxOption.prototype.ngOnDestroy = function () {
        this.destroyed = true;
        clearTimeout(this.scrollTimer);
    };
    NglComboboxOption.ctorParameters = function () { return [
        { type: ElementRef },
        { type: NglComboboxService },
        { type: ChangeDetectorRef },
        { type: NgZone },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglComboboxOption.prototype, "value", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglComboboxOption.prototype, "label", void 0);
    __decorate([
        Input(), InputBoolean(),
        __metadata("design:type", Boolean)
    ], NglComboboxOption.prototype, "selected", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglComboboxOption.prototype, "disabled", void 0);
    __decorate([
        HostListener('mousedown', ['$event']),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [MouseEvent]),
        __metadata("design:returntype", void 0)
    ], NglComboboxOption.prototype, "onSelectViaInteraction", null);
    __decorate([
        HostListener('mouseenter'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], NglComboboxOption.prototype, "hover", null);
    NglComboboxOption = __decorate([
        Component({
            selector: 'ngl-combobox-option, [nglComboboxOption]',
            template: "\n<div class=\"slds-media slds-listbox__option slds-listbox__option_plain slds-media_small slds-media_center\" role=\"option\" [attr.id]=\"uid\" [class.slds-has-focus]=\"active\" [class.slds-is-selected]=\"selected\" [attr.aria-selected]=\"selected || null\" [attr.aria-disabled]=\"disabled || null\"><span class=\"slds-media__figure slds-listbox__option-icon\"><span class=\"slds-icon_container slds-icon-utility-check slds-current-color\" *ngIf=\"selected\">\n      <svg class=\"slds-icon slds-icon_x-small\" nglIconName=\"utility:check\"></svg></span></span><span class=\"slds-media__body\"><span class=\"slds-truncate\"><span class=\"slds-assistive-text\" *ngIf=\"selected\">Current Selection:</span>{{ label }}</span></span></div>",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        __metadata("design:paramtypes", [ElementRef,
            NglComboboxService,
            ChangeDetectorRef,
            NgZone,
            Renderer2])
    ], NglComboboxOption);
    return NglComboboxOption;
}());
export { NglComboboxOption };
