// ng-lightning/esm5/lib/util/convert.js
import { coerceBooleanProperty, _isNumberValue } from '@angular/cdk/coercion';
function propDecoratorFactory(name, fallback) {
    function propDecorator(target, propName) {
        var privatePropName = "$$__" + propName;
        if (Object.prototype.hasOwnProperty.call(target, privatePropName)) {
            console.warn("The prop \"" + privatePropName + "\" is already exist, it will be overrided by " + name + " decorator.");
        }
        Object.defineProperty(target, privatePropName, {
            configurable: true,
            writable: true
        });
        Object.defineProperty(target, propName, {
            get: function () {
                return this[privatePropName];
            },
            set: function (value) {
                this[privatePropName] = fallback(value);
            }
        });
    }
    return propDecorator;
}
export function toBoolean(value) {
    return coerceBooleanProperty(value);
}
export function InputBoolean() {
    return propDecoratorFactory('InputBoolean', toBoolean);
}
export function toNumber(value, fallbackValue) {
    if (fallbackValue === void 0) { fallbackValue = 0; }
    return _isNumberValue(value) ? Number(value) : fallbackValue;
}
export function InputNumber() {
    return propDecoratorFactory('InputNumber', toNumber);
}
