// ng-lightning/esm2015/lib/pills/pill.js
import { __decorate, __metadata } from "tslib";
import { Component, ChangeDetectionStrategy, Input, Output, EventEmitter, HostBinding, } from '@angular/core';
import { InputBoolean } from '../util/convert';
import { isTemplateRef } from '../util/check';
let NglPill = class NglPill {
    constructor() {
        this.isTemplateRef = isTemplateRef;
        /**
           * Applies the error style to the component.
           */
        this.hasError = false;
        /**
           * Whether or not to override the remove button's visibility, if `remove` is set.
           */
        this.removable = true;
        /**
           * Remove button title (and assistive text).
           */
        this.removeTitle = 'Remove';
        /**
           * The event emitted when the remove button is clicked.
           */
        this.remove = new EventEmitter();
        this.linked = false;
    }
    ngOnInit() {
        this.canRemove = this.remove.observers.length > 0;
    }
    onRemove(e) {
        this.remove.emit(e);
    }
    get pillIcon() {
        return this.icon || this.avatar;
    }
};
__decorate([
    Input(),
    __metadata("design:type", Object)
], NglPill.prototype, "icon", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], NglPill.prototype, "avatar", void 0);
__decorate([
    Input(), InputBoolean(), HostBinding('class.slds-has-error'),
    __metadata("design:type", Object)
], NglPill.prototype, "hasError", void 0);
__decorate([
    Input(), InputBoolean(),
    __metadata("design:type", Object)
], NglPill.prototype, "removable", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], NglPill.prototype, "removeTitle", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NglPill.prototype, "remove", void 0);
__decorate([
    HostBinding('class.slds-pill_link'),
    __metadata("design:type", Object)
], NglPill.prototype, "linked", void 0);
NglPill = __decorate([
    Component({
        selector: 'ngl-pill',
        template: "<span class=\"slds-pill__icon_container\" *ngIf=\"pillIcon\">\n  <ng-container *ngIf=\"isTemplateRef(pillIcon); else defaultTpl\" [ngTemplateOutlet]=\"pillIcon\"></ng-container>\n  <ng-template #defaultTpl>\n    <ngl-icon *ngIf=\"icon; else avatarTpl\" [iconName]=\"icon\"></ngl-icon>\n    <ng-template #avatarTpl>\n      <ngl-avatar [src]=\"avatar\" variant=\"circle\"></ngl-avatar>\n    </ng-template>\n  </ng-template></span>\n<ng-container *ngIf=\"linked; else unlinked\">\n  <ng-content select=\"a\"></ng-content>\n</ng-container>\n<ng-template #unlinked><span class=\"slds-pill__label\">\n    <ng-content></ng-content></span></ng-template>\n<button class=\"slds-button slds-button_icon slds-pill__remove\" *ngIf=\"canRemove &amp;&amp; removable\" type=\"button\" [title]=\"removeTitle\" (click)=\"onRemove($event)\">\n  <svg class=\"slds-button__icon\" nglIconName=\"close\"></svg><span class=\"slds-assistive-text\" *ngIf=\"removeTitle\">{{removeTitle}}</span>\n</button>",
        changeDetection: ChangeDetectionStrategy.OnPush,
        host: {
            '[class.slds-pill]': 'true',
        }
    })
], NglPill);
export { NglPill };
