// ng-lightning/esm2015/lib/sections/section.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, Output, EventEmitter, HostBinding, ChangeDetectionStrategy } from '@angular/core';
import { uniqueId } from '../util/util';
let NglExpandableSection = class NglExpandableSection {
    constructor() {
        this.collapsable = true;
        this.open = false;
        this.openChange = new EventEmitter();
        this._uid = uniqueId('expandable-section');
    }
    get expanded() {
        return this.collapsable ? this.open : true;
    }
    get uid() {
        return this.collapsable ? this._uid : undefined;
    }
    toggle(event) {
        event.preventDefault();
        this.openChange.emit(!this.open);
    }
};
__decorate([
    Input(),
    __metadata("design:type", String)
], NglExpandableSection.prototype, "title", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], NglExpandableSection.prototype, "collapsable", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], NglExpandableSection.prototype, "open", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NglExpandableSection.prototype, "openChange", void 0);
__decorate([
    HostBinding('class.slds-is-open'),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], NglExpandableSection.prototype, "expanded", null);
NglExpandableSection = __decorate([
    Component({
        selector: 'ngl-expandable-section',
        changeDetection: ChangeDetectionStrategy.OnPush,
        template: "\n<h3 class=\"slds-section__title\" [class.slds-theme_shade]=\"!collapsable\">\n  <button class=\"slds-button slds-section__title-action\" *ngIf=\"collapsable; else simple_title\" [attr.aria-controls]=\"uid\" [attr.aria-expanded]=\"expanded\" type=\"button\" (click)=\"toggle($event)\">\n    <svg class=\"slds-section__title-action-icon slds-button__icon slds-button__icon_left\" nglIconName=\"switch\"></svg><span class=\"slds-truncate\" [title]=\"title\">{{title}}</span>\n  </button>\n  <ng-template #simple_title><span class=\"slds-truncate slds-p-horizontal_small\" [title]=\"title\">{{title}}</span>\n  </ng-template>\n</h3>\n<div class=\"slds-section__content\" [attr.aria-hidden]=\"!expanded\" [attr.id]=\"uid\">\n  <ng-content></ng-content>\n</div>",
        host: {
            '[class.slds-section]': 'true',
        }
    })
], NglExpandableSection);
export { NglExpandableSection };
