// ng-lightning/esm2015/lib/common/clickoutside.js
import { __decorate, __metadata, __param } from "tslib";
import { Directive, Output, EventEmitter, AfterViewInit, OnDestroy, Inject, ElementRef, Input } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { fromEvent } from 'rxjs';
let NglClickOutsideDirective = class NglClickOutsideDirective {
    constructor(document, element) {
        this.document = document;
        this.element = element;
        this.clickOutside = new EventEmitter();
    }
    ngAfterViewInit() {
        this.subscription = fromEvent(this.document, 'click').subscribe((e) => {
            if (this.shouldClose(e)) {
                this.clickOutside.emit();
            }
        });
    }
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
            this.subscription = null;
        }
    }
    shouldClose(event) {
        const element = event.target;
        if ((event instanceof MouseEvent && event.button === 2) || isContainedIn(element, this.ignore)) {
            return false;
        }
        return !isContainedIn(element, this.element.nativeElement);
    }
};
NglClickOutsideDirective.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: ElementRef }
];
__decorate([
    Output('nglClickOutside'),
    __metadata("design:type", EventEmitter)
], NglClickOutsideDirective.prototype, "clickOutside", void 0);
__decorate([
    Input('nglClickOutsideIgnore'),
    __metadata("design:type", Object)
], NglClickOutsideDirective.prototype, "ignore", void 0);
NglClickOutsideDirective = __decorate([
    Directive({
        selector: '[nglClickOutside]'
    }),
    __param(0, Inject(DOCUMENT)),
    __metadata("design:paramtypes", [Object, ElementRef])
], NglClickOutsideDirective);
export { NglClickOutsideDirective };
function isContainedIn(el, container) {
    if (!container) {
        return false;
    }
    return Array.isArray(container) ? container.some(c => c.contains(el)) : container.contains(el);
}
