// ng-lightning/esm2015/lib/avatar/avatar.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, Output, EventEmitter, ChangeDetectionStrategy, ElementRef, Renderer2, HostBinding, OnInit, OnChanges } from '@angular/core';
import { HostService } from '../common/host/host.service';
let NglAvatar = class NglAvatar {
    constructor(element, renderer, hostService) {
        this.element = element;
        this.hostService = hostService;
        this.src = '';
        this.alternativeText = '';
        this.fallbackIconName = 'standard:user';
        this.error = new EventEmitter();
        this._imgError = false;
        renderer.addClass(element.nativeElement, 'slds-avatar');
    }
    fallbackIconClass() {
        const [category, icon] = this.fallbackIconName.split(':');
        return `slds-icon-${category}-${icon}`;
    }
    get shouldShowImage() {
        return this.src && !this._imgError;
    }
    onImgError() {
        this._imgError = true;
        this.error.emit();
    }
    ngOnInit() {
        this.setHostClass();
    }
    ngOnChanges() {
        this.setHostClass();
    }
    setHostClass() {
        this.hostService.updateClass(this.element, {
            [`slds-avatar_${this.size || 'medium'}`]: true,
            [`slds-avatar_${this.variant || 'rectangle'}`]: true,
        });
    }
};
NglAvatar.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: HostService }
];
__decorate([
    Input(),
    __metadata("design:type", Object)
], NglAvatar.prototype, "src", void 0);
__decorate([
    HostBinding('attr.title'),
    Input(),
    __metadata("design:type", Object)
], NglAvatar.prototype, "alternativeText", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], NglAvatar.prototype, "size", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], NglAvatar.prototype, "variant", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], NglAvatar.prototype, "initials", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], NglAvatar.prototype, "fallbackIconName", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NglAvatar.prototype, "error", void 0);
NglAvatar = __decorate([
    Component({
        selector: 'ngl-avatar',
        changeDetection: ChangeDetectionStrategy.OnPush,
        template: "<img *ngIf=\"shouldShowImage; else template_initials\" [src]=\"src\" [alt]=\"alternativeText\" (error)=\"onImgError()\">\n<ng-template #template_initials><abbr class=\"slds-avatar__initials\" [ngClass]=\"fallbackIconClass()\">{{ initials }}</abbr></ng-template>",
        providers: [HostService]
    }),
    __metadata("design:paramtypes", [ElementRef, Renderer2, HostService])
], NglAvatar);
export { NglAvatar };
