// ng-lightning/esm5/lib/colorpicker/swatch/colorpicker-swatch.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ElementRef, Renderer2, ChangeDetectionStrategy, HostBinding } from '@angular/core';
var NglColorpickerSwatch = /** @class */ (function () {
    function NglColorpickerSwatch(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.renderer.addClass(this.el.nativeElement, 'slds-swatch');
    }
    NglColorpickerSwatch.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        HostBinding('style.background'),
        Input(),
        __metadata("design:type", String)
    ], NglColorpickerSwatch.prototype, "color", void 0);
    NglColorpickerSwatch = __decorate([
        Component({
            // tslint:disable-next-line:component-selector
            selector: '[nglColorpickerSwatch]',
            template: "<span class=\"slds-assistive-text\" aria-hidden=\"true\">{{ color }}</span>",
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2])
    ], NglColorpickerSwatch);
    return NglColorpickerSwatch;
}());
export { NglColorpickerSwatch };
