// ng-lightning/esm5/lib/files/file-crop.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, Renderer2, Input } from '@angular/core';
var NglFileCrop = /** @class */ (function () {
    function NglFileCrop(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.cropClass = 'slds-file__crop';
        // this.renderer.addClass(this.element.nativeElement, this.cropClass);
    }
    Object.defineProperty(NglFileCrop.prototype, "nglFileCrop", {
        set: function (ratio) {
            var nativeElement = this.element.nativeElement;
            if (this.currentRatio) {
                this.renderer.removeClass(nativeElement, "" + this.cropClass);
                this.renderer.removeClass(nativeElement, this.cropClass + "_" + this.currentRatio);
            }
            if (ratio) {
                this.renderer.addClass(nativeElement, "" + this.cropClass);
                this.renderer.addClass(nativeElement, this.cropClass + "_" + ratio);
            }
            this.currentRatio = ratio;
        },
        enumerable: true,
        configurable: true
    });
    NglFileCrop.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], NglFileCrop.prototype, "nglFileCrop", null);
    NglFileCrop = __decorate([
        Directive({
            selector: '[nglFileCrop]',
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2])
    ], NglFileCrop);
    return NglFileCrop;
}());
export { NglFileCrop };
