// ng-lightning/esm2015/lib/menus/dropdown-item.js
import { __decorate, __metadata } from "tslib";
import { Directive, ElementRef, HostListener } from '@angular/core';
let NglDropdownItem = class NglDropdownItem {
    constructor(element) {
        this.element = element;
        this.isFocused = false;
    }
    onFocus() {
        this.isFocused = true;
    }
    onBlur() {
        this.isFocused = false;
    }
    hasFocus() {
        return this.isFocused;
    }
    focus() {
        this.element.nativeElement.focus();
    }
};
NglDropdownItem.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    HostListener('focus'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], NglDropdownItem.prototype, "onFocus", null);
__decorate([
    HostListener('blur'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], NglDropdownItem.prototype, "onBlur", null);
NglDropdownItem = __decorate([
    Directive({
        selector: '[nglDropdownItem]',
        host: {
            'tabindex': '0',
        },
    }),
    __metadata("design:paramtypes", [ElementRef])
], NglDropdownItem);
export { NglDropdownItem };
