// ng-lightning/esm2015/lib/menus/dropdown-trigger.js
import { __decorate, __metadata } from "tslib";
import { Directive, HostListener, ElementRef, OnDestroy } from '@angular/core';
import { NglDropdown } from './dropdown';
let NglDropdownTrigger = class NglDropdownTrigger {
    constructor(element, dropdown) {
        this.element = element;
        this.dropdown = dropdown;
        this.parentFocusEventSubscription = this.dropdown.triggerFocusEventEmitter.subscribe(this.focus.bind(this));
    }
    ngOnDestroy() {
        this.parentFocusEventSubscription.unsubscribe();
    }
    toggleOpen() {
        this.dropdown.toggle();
    }
    onKeyDownOpen($event) {
        $event.preventDefault();
        this.dropdown.toggle(true);
    }
    focus() {
        this.element.nativeElement.focus();
    }
};
NglDropdownTrigger.ctorParameters = () => [
    { type: ElementRef },
    { type: NglDropdown }
];
__decorate([
    HostListener('click'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], NglDropdownTrigger.prototype, "toggleOpen", null);
__decorate([
    HostListener('keydown.arrowdown', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Event]),
    __metadata("design:returntype", void 0)
], NglDropdownTrigger.prototype, "onKeyDownOpen", null);
NglDropdownTrigger = __decorate([
    Directive({
        selector: '[nglDropdownTrigger]',
        host: {
            'aria-haspopup': 'true',
        },
    }),
    __metadata("design:paramtypes", [ElementRef, NglDropdown])
], NglDropdownTrigger);
export { NglDropdownTrigger };
