// ng-lightning/esm5/lib/ng-lightning.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { NglAccordionModule } from './accordion/module';
import { NglAlertModule } from './alert/module';
import { NglAvatarModule } from './avatar/module';
import { NglBadgesModule } from './badges/module';
import { NglBreadcrumbsModule } from './breadcrumbs/module';
import { NglButtonIconsModule } from './button-icons/module';
import { NglButtonsModule } from './buttons/module';
import { NglCarouselModule } from './carousel/module';
import { NglCheckboxesModule } from './checkboxes/module';
import { NglColorpickerModule } from './colorpicker/module';
import { NglComboboxesModule } from './comboboxes/module';
import { NglDatatablesModule } from './datatables/module';
import { NglDatepickersModule } from './datepickers/module';
import { NglDynamicIconsModule } from './dynamicicons/module';
import { NglFilesModule } from './files/module';
import { NglFileUploadModule } from './file-upload/module';
import { NglIconsModule } from './icons/module';
import { NglInputModule } from './input/module';
import { NglMenusModule } from './menus/module';
import { NglModalsModule } from './modals/module';
import { NglToastModule } from './toast/module';
import { NglPaginationsModule } from './paginations/module';
import { NglPickModule } from './pick/module';
import { NglPillsModule } from './pills/module';
import { NglPopoversModule } from './popovers/module';
import { NglProgressBarModule } from './progressbar/module';
import { NglRadiosModule } from './radio-group/module';
import { NglRatingsModule } from './ratings/module';
import { NglSectionsModule } from './sections/module';
import { NglSelectModule } from './select/module';
import { NglSliderModule } from './slider/module';
import { NglSpinnersModule } from './spinners/module';
import { NglTabsModule } from './tabs/module';
import { NglTextareaModule } from './textarea/module';
import { NglTooltipsModule } from './tooltips/module';
export { NglAccordionModule } from './accordion/module';
export { NglAlertModule } from './alert/module';
export { NglAvatarModule } from './avatar/module';
export { NglBadgesModule } from './badges/module';
export { NglBreadcrumbsModule } from './breadcrumbs/module';
export { NglButtonIconsModule } from './button-icons/module';
export { NglButtonsModule } from './buttons/module';
export { NglCarouselModule } from './carousel/module';
export { NglCheckboxesModule } from './checkboxes/module';
export { NglColorpickerModule } from './colorpicker/module';
export { NglComboboxesModule } from './comboboxes/module';
export { NglDatatablesModule } from './datatables/module';
export { NglDatepickersModule } from './datepickers/module';
export { NglDynamicIconsModule } from './dynamicicons/module';
export { NglFilesModule } from './files/module';
export { NglFileUploadModule } from './file-upload/module';
export { NglIconsModule } from './icons/module';
export { NglInputModule } from './input/module';
export { NglMenusModule } from './menus/module';
export { NglModalsModule } from './modals/module';
export { NglToastModule } from './toast/module';
export { NglPaginationsModule } from './paginations/module';
export { NglPickModule } from './pick/module';
export { NglPillsModule } from './pills/module';
export { NglPopoversModule } from './popovers/module';
export { NglProgressBarModule } from './progressbar/module';
export { NglRadiosModule } from './radio-group/module';
export { NglRatingsModule } from './ratings/module';
export { NglSectionsModule } from './sections/module';
export { NglSelectModule } from './select/module';
export { NglSliderModule } from './slider/module';
export { NglSpinnersModule } from './spinners/module';
export { NglTabsModule } from './tabs/module';
export { NglTextareaModule } from './textarea/module';
export { NglTooltipsModule } from './tooltips/module';
var MODULES = [
    NglAccordionModule,
    NglAlertModule,
    NglAvatarModule,
    NglBadgesModule,
    NglBreadcrumbsModule,
    NglButtonIconsModule,
    NglButtonsModule,
    NglCarouselModule,
    NglCheckboxesModule,
    NglColorpickerModule,
    NglComboboxesModule,
    NglDatatablesModule,
    NglDatepickersModule,
    NglDynamicIconsModule,
    NglFilesModule,
    NglFileUploadModule,
    NglIconsModule,
    NglInputModule,
    NglMenusModule,
    NglModalsModule,
    NglToastModule,
    NglPaginationsModule,
    NglPickModule,
    NglPillsModule,
    NglPopoversModule,
    NglProgressBarModule,
    NglRadiosModule,
    NglRatingsModule,
    NglSectionsModule,
    NglSelectModule,
    NglSliderModule,
    NglSpinnersModule,
    NglTabsModule,
    NglTextareaModule,
    NglTooltipsModule,
];
var NglModule = /** @class */ (function () {
    function NglModule() {
    }
    NglModule = __decorate([
        NgModule({
            exports: MODULES,
        })
    ], NglModule);
    return NglModule;
}());
export { NglModule };
