// ng-lightning/esm5/lib/common/notify/notify.js
import { __decorate, __metadata } from "tslib";
import { Input, Output, EventEmitter } from '@angular/core';
import { isInt } from '../../util/util';
var NglCommonNotify = /** @class */ (function () {
    function NglCommonNotify(element, renderer, cd, type) {
        this.element = element;
        this.renderer = renderer;
        this.cd = cd;
        this.closeButtonAssistiveText = 'Close';
        /**
         * Triggered by close button or duration timeout.
         */
        // tslint:disable-next-line:no-output-rename
        this.closeEventEmitter = new EventEmitter();
        this.currentTimeout = null;
        this.renderer.addClass(this.element.nativeElement, 'slds-notify');
        this.renderer.addClass(this.element.nativeElement, "slds-notify_" + type);
        this.toggleThemeClass(true, this.variant);
        this.renderer.setAttribute(this.element.nativeElement, 'role', type === 'toast' ? 'status' : 'alert');
    }
    Object.defineProperty(NglCommonNotify.prototype, "variant", {
        get: function () {
            return this._variant || 'info';
        },
        /**
         * The type of alert.
         */
        set: function (variant) {
            this.toggleThemeClass(false, this.variant);
            this._variant = variant;
            this.toggleThemeClass(true, this.variant);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NglCommonNotify.prototype, "duration", {
        /**
         * The number of milliseconds after which, the close event will be triggered with an emitted reason of `'timeout'`.
         */
        set: function (duration) {
            var _this = this;
            this.clearTimeout();
            if (isInt(duration) && duration >= 0) {
                this.currentTimeout = setTimeout(function () { return _this.close('timeout'); }, +duration);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NglCommonNotify.prototype, "dismissible", {
        get: function () {
            return this._dismissible;
        },
        set: function (dismissible) {
            this._dismissible = dismissible;
            this.cd.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    NglCommonNotify.prototype.close = function (reason, $event) {
        this.clearTimeout();
        if ($event) {
            $event.preventDefault();
            $event.stopPropagation();
        }
        this.closeEventEmitter.emit(reason);
    };
    NglCommonNotify.prototype.ngOnDestroy = function () {
        this.clearTimeout();
    };
    NglCommonNotify.prototype.clearTimeout = function () {
        if (this.currentTimeout !== null) {
            clearTimeout(this.currentTimeout);
            this.currentTimeout = null;
        }
    };
    NglCommonNotify.prototype.toggleThemeClass = function (isAdd, klass) {
        if (!klass) {
            return;
        }
        var el = this.element.nativeElement;
        this.renderer[isAdd ? 'addClass' : 'removeClass'](el, "slds-theme_" + klass);
    };
    __decorate([
        Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], NglCommonNotify.prototype, "variant", null);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglCommonNotify.prototype, "iconName", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglCommonNotify.prototype, "assistiveText", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglCommonNotify.prototype, "closeButtonAssistiveText", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], NglCommonNotify.prototype, "duration", null);
    __decorate([
        Output('close'),
        __metadata("design:type", Object)
    ], NglCommonNotify.prototype, "closeEventEmitter", void 0);
    return NglCommonNotify;
}());
export { NglCommonNotify };
