// ng-lightning/esm5/lib/datepickers/util.js
export function parseDate(date) {
    if (!date) {
        return null;
    }
    return { year: date.getFullYear(), month: date.getMonth(), day: date.getDate() };
}
export function isEqualDate(d1, d2) {
    return d1 && d2 && d1.day === d2.day && d1.month === d2.month && d1.year === d2.year;
}
export function getToday() {
    var today = new Date();
    return { year: today.getFullYear(), month: today.getMonth(), day: today.getDate() };
}
export function numberOfDaysInMonth(year, month) {
    return new Date(year, month + 1, 0).getDate();
}
// Split array into smaller arrays
export function split(arr, size) {
    if (size === void 0) { size = 7; }
    var arrays = [];
    while (arr.length > 0) {
        arrays.push(arr.splice(0, size));
    }
    return arrays;
}
export function isDisabled(d, disabledCallback, min, max) {
    var date = new Date(d.year, d.month, d.day);
    return (disabledCallback && disabledCallback(date)) ||
        (min && compareDate(d, min) < 0) ||
        (max && compareDate(d, max) > 0);
}
export function compareDate(d1, d2) {
    if (isEqualDate(d1, d2)) {
        return 0;
    }
    var keys = ['year', 'month', 'day'];
    for (var i = 0; i < 3; i++) {
        var key = keys[i];
        var diff = d1[key] - d2[key];
        if (diff !== 0) {
            return diff > 0 ? 1 : -1;
        }
    }
}
export function isSameMonth(d1, d2) {
    return d1.year === d2.year && d1.month === d2.month;
}
