// ng-lightning/esm5/lib/datepickers/weekdays.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ChangeDetectionStrategy } from '@angular/core';
var NglDatepickerWeekdays = /** @class */ (function () {
    function NglDatepickerWeekdays() {
        this.weekdays = [];
    }
    NglDatepickerWeekdays.prototype.ngOnChanges = function (changes) {
        this.weekdays = [];
        for (var i = 0; i < 7; i++) {
            var offset = (this.firstDayOfWeek + i) % 7;
            this.weekdays.push({
                id: "weekday-" + i,
                label: this.dayNamesShort[offset],
                title: this.dayNamesLong[offset],
            });
        }
    };
    __decorate([
        Input(),
        __metadata("design:type", Array)
    ], NglDatepickerWeekdays.prototype, "dayNamesShort", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Array)
    ], NglDatepickerWeekdays.prototype, "dayNamesLong", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], NglDatepickerWeekdays.prototype, "firstDayOfWeek", void 0);
    NglDatepickerWeekdays = __decorate([
        Component({
            // tslint:disable-next-line:component-selector
            selector: 'tr[nglWeekdays]',
            template: "\n<th *ngFor=\"let day of weekdays\" [id]=\"day.id\" scope=\"col\"><abbr [title]=\"day.title\">{{day.label}}</abbr></th>",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], NglDatepickerWeekdays);
    return NglDatepickerWeekdays;
}());
export { NglDatepickerWeekdays };
