// ng-lightning/esm5/lib/datepickers/year.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, Output, EventEmitter, ChangeDetectionStrategy } from '@angular/core';
import { uniqueId } from '../util/util';
import { InputNumber } from '../util/convert';
var NglDatepickerYear = /** @class */ (function () {
    function NglDatepickerYear() {
        this.uid = uniqueId('datepicker_year');
        this.yearChange = new EventEmitter();
    }
    NglDatepickerYear.prototype.change = function ($event) {
        this.yearChange.emit($event);
    };
    NglDatepickerYear.prototype.ngOnChanges = function () {
        this.range = this.getRange();
    };
    NglDatepickerYear.prototype.getRange = function () {
        var minYear = Math.min(this.from.year, this.year);
        var maxYear = Math.max(this.to.year, this.year);
        var size = maxYear - minYear;
        return Array.apply(null, { length: size + 1 }).map(function (value, index) { return minYear + index; });
    };
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglDatepickerYear.prototype, "from", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NglDatepickerYear.prototype, "to", void 0);
    __decorate([
        Input(), InputNumber(),
        __metadata("design:type", Number)
    ], NglDatepickerYear.prototype, "year", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NglDatepickerYear.prototype, "yearChange", void 0);
    NglDatepickerYear = __decorate([
        Component({
            selector: 'ngl-date-year',
            template: "\n<label class=\"slds-assistive-text\" [attr.for]=\"uid\">Pick a Year</label>\n<div class=\"slds-select_container\">\n  <select class=\"slds-select\" [id]=\"uid\" [ngModel]=\"year\" (ngModelChange)=\"change($event)\">\n    <option *ngFor=\"let yr of range\" [value]=\"yr\">{{yr}}</option>\n  </select>\n</div>",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], NglDatepickerYear);
    return NglDatepickerYear;
}());
export { NglDatepickerYear };
