// ng-lightning/esm5/lib/spinners/spinner.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ChangeDetectionStrategy, ElementRef, Renderer2, OnInit, OnChanges } from '@angular/core';
import { HostService } from '../common/host/host.service';
var NglSpinner = /** @class */ (function () {
    function NglSpinner(element, renderer, hostService) {
        this.element = element;
        this.renderer = renderer;
        this.hostService = hostService;
        this.renderer.addClass(this.element.nativeElement, 'slds-spinner');
        this.renderer.setAttribute(this.element.nativeElement, 'role', 'status');
    }
    NglSpinner.prototype.ngOnInit = function () {
        this.setHostClass();
    };
    NglSpinner.prototype.ngOnChanges = function () {
        this.setHostClass();
    };
    NglSpinner.prototype.setHostClass = function () {
        var _a;
        this.hostService.updateClass(this.element, (_a = {},
            _a["slds-spinner_" + (this.size || 'medium')] = true,
            _a["slds-spinner_" + this.variant] = !!this.variant,
            _a));
    };
    NglSpinner.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: HostService }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglSpinner.prototype, "size", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglSpinner.prototype, "variant", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NglSpinner.prototype, "alternativeText", void 0);
    NglSpinner = __decorate([
        Component({
            selector: 'ngl-spinner',
            template: "<span class=\"slds-assistive-text\" *ngIf=\"alternativeText\">{{ alternativeText }}</span>\n<div class=\"slds-spinner__dot-a\"></div>\n<div class=\"slds-spinner__dot-b\"></div>",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [HostService]
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2, HostService])
    ], NglSpinner);
    return NglSpinner;
}());
export { NglSpinner };
