// ng-lightning/esm5/lib/colorpicker/custom/inputs/colorpicker-inputs.js
import { __decorate, __metadata } from "tslib";
import { Component, ChangeDetectionStrategy, Input, Output, EventEmitter } from '@angular/core';
import { _isNumberValue } from '@angular/cdk/coercion';
import { getHexFromRgb, getRgbFromHex, isValidHex } from '../../util';
import { uniqueId } from '../../../util/util';
var NglColorpickerInputs = /** @class */ (function () {
    function NglColorpickerInputs() {
        this.hexChange = new EventEmitter();
        this.red = 0;
        this.green = 0;
        this.blue = 0;
        this.uid = uniqueId('colorpicker-inputs');
    }
    Object.defineProperty(NglColorpickerInputs.prototype, "hex", {
        get: function () {
            return this._hex;
        },
        set: function (hex) {
            if (hex) {
                this._hex = hex;
                var _a = getRgbFromHex(this.hex), red = _a.red, green = _a.green, blue = _a.blue;
                this.red = red;
                this.green = green;
                this.blue = blue;
            }
        },
        enumerable: true,
        configurable: true
    });
    NglColorpickerInputs.prototype.updateHex = function (value) {
        var isValid = isValidHex(value);
        if (!isValid) {
            this.red = this.green = this.blue = null;
        }
        this.hexChange.emit(isValid ? value : null);
    };
    NglColorpickerInputs.prototype.onRGB = function (key, value) {
        this[key] = value;
        var hex = this.isRGBValid() ? getHexFromRgb({ red: this.red, green: this.green, blue: this.blue }) : null;
        this.hexChange.emit(hex);
    };
    Object.defineProperty(NglColorpickerInputs.prototype, "isHexInvalid", {
        get: function () {
            return this.red === null && this.green === null && this.blue === null;
        },
        enumerable: true,
        configurable: true
    });
    NglColorpickerInputs.prototype.isColorNumberValid = function (key) {
        var value = this[key];
        return _isNumberValue(value) && value >= 0 && value <= 255;
    };
    NglColorpickerInputs.prototype.isRGBValid = function () {
        var _this = this;
        return ['red', 'green', 'blue'].every(function (prop) { return _this.isColorNumberValid(prop); });
    };
    __decorate([
        Input(),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [String])
    ], NglColorpickerInputs.prototype, "hex", null);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NglColorpickerInputs.prototype, "hexChange", void 0);
    NglColorpickerInputs = __decorate([
        Component({
            selector: 'ngl-colorpicker-inputs',
            template: "\n<div class=\"slds-color-picker__custom-inputs\">\n  <div class=\"slds-form-element slds-color-picker__input-custom-hex\" [class.slds-has-error]=\"isHexInvalid\">\n    <label class=\"slds-form-element__label\" [attr.for]=\"uid + 'hex'\">Hex</label>\n    <div class=\"slds-form-element__control\">\n      <input class=\"slds-input\" [id]=\"uid + 'hex'\" type=\"text\" maxlength=\"7\" [value]=\"hex\" (input)=\"updateHex($event.target.value)\">\n    </div>\n  </div>\n  <div class=\"slds-form-element\" [class.slds-has-error]=\"!isColorNumberValid('red')\">\n    <label class=\"slds-form-element__label\" [attr.for]=\"uid + 'red'\"><abbr title=\"red\">R</abbr></label>\n    <div class=\"slds-form-element__control\">\n      <input class=\"slds-input\" [id]=\"uid + 'red'\" type=\"text\" maxlength=\"3\" [value]=\"red\" (input)=\"onRGB('red', $event.target.value)\">\n    </div>\n  </div>\n  <div class=\"slds-form-element\" [class.slds-has-error]=\"!isColorNumberValid('green')\">\n    <label class=\"slds-form-element__label\" [attr.for]=\"uid + 'green'\"><abbr title=\"green\">G</abbr></label>\n    <div class=\"slds-form-element__control\">\n      <input class=\"slds-input\" [id]=\"uid + 'green'\" type=\"text\" maxlength=\"3\" [value]=\"green\" (input)=\"onRGB('green', $event.target.value)\">\n    </div>\n  </div>\n  <div class=\"slds-form-element\" [class.slds-has-error]=\"!isColorNumberValid('blue')\">\n    <label class=\"slds-form-element__label\" [attr.for]=\"uid + 'blue'\"><abbr title=\"blue\">B</abbr></label>\n    <div class=\"slds-form-element__control\">\n      <input class=\"slds-input\" [id]=\"uid + 'blue'\" type=\"text\" maxlength=\"3\" [value]=\"blue\" (input)=\"onRGB('blue', $event.target.value)\">\n    </div>\n  </div>\n</div>\n<div class=\"slds-color-picker\">\n  <p class=\"slds-form-error slds-color-picker__input-custom-error\" *ngIf=\"isHexInvalid; else rgbError\">The color entered is invalid</p>\n  <ng-template #rgbError>\n    <p class=\"slds-form-error slds-color-picker__input-custom-error\" *ngIf=\"!isRGBValid()\">The value needs to be an integer from 0-255</p>\n  </ng-template>\n</div>",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], NglColorpickerInputs);
    return NglColorpickerInputs;
}());
export { NglColorpickerInputs };
