// ng-lightning/esm2015/lib/datatables/cell-internal.js
import { __decorate, __metadata } from "tslib";
import { Component, ChangeDetectionStrategy, Input, HostBinding } from '@angular/core';
import { NglDatatableColumn } from './column';
let NglInternalDatatableCell = class NglInternalDatatableCell {
    get dataLabel() {
        return this.column.heading;
    }
    ngOnChanges() {
        this.context = {
            $implicit: this.value,
            row: this.row,
            index: this.index,
        };
    }
    get value() {
        const { key } = this.column;
        return key ? this.row[key] : null;
    }
};
__decorate([
    Input(),
    __metadata("design:type", Object)
], NglInternalDatatableCell.prototype, "row", void 0);
__decorate([
    Input(),
    __metadata("design:type", NglDatatableColumn)
], NglInternalDatatableCell.prototype, "column", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number)
], NglInternalDatatableCell.prototype, "index", void 0);
__decorate([
    HostBinding('attr.data-label'),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], NglInternalDatatableCell.prototype, "dataLabel", null);
NglInternalDatatableCell = __decorate([
    Component({
        // tslint:disable-next-line:component-selector
        selector: 'td[nglDatatatableCell_]',
        template: "\n<div [class.slds-truncate]=\"column.truncate\" [attr.title]=\"column.truncate ? value : null\">\n  <ng-container *ngIf=\"column.cellTpl; else stringTpl\" [ngTemplateOutlet]=\"column.cellTpl.templateRef\" [ngTemplateOutletContext]=\"context\"></ng-container>\n  <ng-template #stringTpl>{{ value }}</ng-template>\n</div>",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], NglInternalDatatableCell);
export { NglInternalDatatableCell };
