// ng-lightning/esm2015/lib/datatables/datatable.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ChangeDetectorRef, ContentChild, ContentChildren, QueryList, HostBinding, Output, EventEmitter, AfterContentInit, OnDestroy } from '@angular/core';
import { NglDatatableColumn } from './column';
import { NglDatatableLoadingOverlay, NglDatatableNoRowsOverlay } from './overlays';
let NglDatatable = class NglDatatable {
    constructor(detector) {
        this.detector = detector;
        this.data = [];
        this.sortChange = new EventEmitter();
        this.loading = false;
        this.rowClick = new EventEmitter();
        this.dataTrackBy = (index, data) => {
            return this.trackByKey ? data[this.trackByKey] : index;
        };
    }
    get showLoading() {
        return this.loading && this.loadingOverlay;
    }
    columnTrackBy(index, column) {
        return column.key || index;
    }
    onColumnSort(column, order) {
        const key = column.key;
        if (!key) {
            throw new Error(`ng-lightning: No "key" property is set for sortable column "${column.heading}"`);
        }
        this.sortChange.emit({ key, order });
    }
    getColumnSortOrder(column) {
        return this.sort && column.key === this.sort.key ? this.sort.order : null;
    }
    onRowClick(event, data) {
        this.rowClick.emit({ event, data });
    }
    ngAfterContentInit() {
        this._columnsSubscription = this.columns.changes.subscribe(() => this.detector.markForCheck());
    }
    ngOnDestroy() {
        if (this._columnsSubscription) {
            this._columnsSubscription.unsubscribe();
            this._columnsSubscription = null;
        }
    }
};
NglDatatable.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
__decorate([
    Input(),
    __metadata("design:type", Array)
], NglDatatable.prototype, "data", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], NglDatatable.prototype, "trackByKey", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], NglDatatable.prototype, "sort", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NglDatatable.prototype, "sortChange", void 0);
__decorate([
    HostBinding('class.slds-is-relative'),
    Input(),
    __metadata("design:type", Object)
], NglDatatable.prototype, "loading", void 0);
__decorate([
    ContentChild(NglDatatableLoadingOverlay),
    __metadata("design:type", NglDatatableLoadingOverlay)
], NglDatatable.prototype, "loadingOverlay", void 0);
__decorate([
    ContentChild(NglDatatableNoRowsOverlay),
    __metadata("design:type", NglDatatableNoRowsOverlay)
], NglDatatable.prototype, "noRowsOverlay", void 0);
__decorate([
    ContentChildren(NglDatatableColumn),
    __metadata("design:type", QueryList)
], NglDatatable.prototype, "columns", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NglDatatable.prototype, "rowClick", void 0);
NglDatatable = __decorate([
    Component({
        // tslint:disable-next-line:component-selector
        selector: 'table[ngl-datatable]',
        template: "\n<thead>\n  <tr class=\"slds-line-height_reset\">\n    <th *ngFor=\"let col of columns; trackBy:columnTrackBy\" nglDatatableHead scope=\"col\" [heading]=\"col.heading\" [headingTpl]=\"col.headingTpl?.templateRef\" [sortable]=\"col.sortable\" [sortOrder]=\"getColumnSortOrder(col)\" (sort)=\"onColumnSort(col, $event)\" [ngClass]=\"col.headClass\"></th>\n  </tr>\n</thead>\n<tbody>\n  <ng-template #noData>\n    <tr>\n      <td [attr.colspan]=\"columns.length\">\n        <ng-template [ngTemplateOutlet]=\"noRowsOverlay?.templateRef\"></ng-template>\n      </td>\n    </tr>\n  </ng-template>\n  <ng-container *ngIf=\"data &amp;&amp; data.length &gt; 0; else noData\">\n    <tr *ngFor=\"let d of data; let i = index; trackBy:dataTrackBy\" (click)=\"onRowClick($event, d)\">\n      <td *ngFor=\"let col of columns; trackBy:columnTrackBy\" [ngClass]=\"col.cellClass\" nglDatatatableCell_ [row]=\"d\" [column]=\"col\" [index]=\"i\"></td>\n    </tr>\n  </ng-container>\n</tbody>\n<div class=\"ngl-datatable-loading slds-align_absolute-center\" *ngIf=\"showLoading\">\n  <ng-template [ngTemplateOutlet]=\"loadingOverlay.templateRef\"></ng-template>\n</div>",
        host: {
            '[class.slds-table]': 'true',
        },
        styles: [`
    .ngl-datatable-loading {
      position: absolute;
      z-index: 1;
      top: 0; left: 0; right: 0; bottom: 0;
      background: rgba(255, 255, 255, 0.5)
    }
  `]
    }),
    __metadata("design:paramtypes", [ChangeDetectorRef])
], NglDatatable);
export { NglDatatable };
