// ng-lightning/esm2015/lib/datatables/head.js
import { __decorate, __metadata } from "tslib";
import { Component, ChangeDetectionStrategy, Input, Output, EventEmitter, HostBinding, TemplateRef } from '@angular/core';
let NglInternalDatatableHeadCell = class NglInternalDatatableHeadCell {
    constructor() {
        this.sort = new EventEmitter();
    }
    get header() {
        return this.headingTpl || this.heading;
    }
    get attrTitle() {
        return this.heading || null;
    }
    get ariaSort() {
        return this.sortOrder ? `${this.sortOrder}ending` : 'none';
    }
    sortChange() {
        this.sort.emit(this.sortOrder === 'desc' ? 'asc' : 'desc');
    }
};
__decorate([
    Input(),
    __metadata("design:type", String)
], NglInternalDatatableHeadCell.prototype, "heading", void 0);
__decorate([
    Input(),
    __metadata("design:type", TemplateRef)
], NglInternalDatatableHeadCell.prototype, "headingTpl", void 0);
__decorate([
    HostBinding('class.slds-is-sortable'),
    Input(),
    __metadata("design:type", Boolean)
], NglInternalDatatableHeadCell.prototype, "sortable", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], NglInternalDatatableHeadCell.prototype, "sortOrder", void 0);
__decorate([
    HostBinding('attr.aria-sort'),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], NglInternalDatatableHeadCell.prototype, "ariaSort", null);
__decorate([
    Output(),
    __metadata("design:type", Object)
], NglInternalDatatableHeadCell.prototype, "sort", void 0);
NglInternalDatatableHeadCell = __decorate([
    Component({
        // tslint:disable-next-line:component-selector
        selector: 'th[nglDatatableHead]',
        template: "<a class=\"slds-th__action slds-text-link_reset\" *ngIf=\"sortable; else baseTpl\" (click)=\"sortChange()\" role=\"button\" tabindex=\"0\"><span class=\"slds-assistive-text\">Sort by:</span>\n  <div class=\"slds-grid slds-grid_vertical-align-center slds-has-flexi-truncate\"><span class=\"slds-truncate\" [attr.title]=\"attrTitle\" [nglInternalOutlet]=\"header\"></span><span class=\"slds-icon_container slds-icon-utility-arrowdown\">\n      <svg class=\"slds-icon slds-icon-text-default slds-is-sortable__icon\" nglIconName=\"arrowdown\"></svg></span></div></a>\n<ng-template #baseTpl>\n  <div class=\"slds-truncate\" [attr.title]=\"attrTitle\" [nglInternalOutlet]=\"header\"></div>\n</ng-template>",
        changeDetection: ChangeDetectionStrategy.OnPush,
        host: {
            '[class.slds-is-sorted_asc]': `sortOrder === 'asc'`,
            '[class.slds-is-sorted_desc]': `sortOrder === 'desc'`,
            '[class.slds-is-sorted]': `!!sortOrder`,
        }
    })
], NglInternalDatatableHeadCell);
export { NglInternalDatatableHeadCell };
