// ng-lightning/esm5/lib/common/host/host.service.js
import { __assign, __decorate, __metadata } from "tslib";
import { Injectable, Renderer2, RendererFactory2, ElementRef } from '@angular/core';
var HostService = /** @class */ (function () {
    function HostService(rendererFactory2) {
        this.classMap = {};
        this.styleMap = {};
        this.renderer = rendererFactory2.createRenderer(null, null);
    }
    HostService.prototype.updateClass = function (_a, classMap) {
        var _this = this;
        var nativeElement = _a.nativeElement;
        var newClassMap = {};
        var remove = __assign({}, this.classMap);
        Object.keys(classMap).filter(function (i) { return classMap[i]; }).forEach(function (i) {
            newClassMap[i] = true;
            if (!_this.classMap[i]) {
                _this.renderer.addClass(nativeElement, i);
            }
            if (remove[i]) {
                remove[i] = false;
            }
        });
        Object.keys(remove).filter(function (i) { return remove[i]; }).forEach(function (i) { return _this.renderer.removeClass(nativeElement, i); });
        this.classMap = newClassMap;
    };
    HostService.prototype.updateStyle = function (_a, styleMap) {
        var _this = this;
        var nativeElement = _a.nativeElement;
        var remove = __assign({}, this.styleMap);
        Object.keys(styleMap).filter(function (i) { return styleMap[i]; }).forEach(function (i) {
            if (styleMap[i] !== false) {
                _this.renderer.setStyle(nativeElement, i, styleMap[i]);
            }
            if (remove[i]) {
                delete remove[i];
            }
        });
        Object.keys(remove).forEach(function (i) { return _this.renderer.removeStyle(nativeElement, i); });
        this.styleMap = styleMap;
    };
    HostService.ctorParameters = function () { return [
        { type: RendererFactory2 }
    ]; };
    HostService = __decorate([
        Injectable(),
        __metadata("design:paramtypes", [RendererFactory2])
    ], HostService);
    return HostService;
}());
export { HostService };
