// ng-lightning/esm2015/lib/carousel/carousel-indicator.js
import { __decorate, __metadata } from "tslib";
import { Directive, HostBinding, Input, ElementRef, OnChanges, SimpleChanges } from '@angular/core';
import { InputBoolean } from '../util/convert';
import { NglCarouselImage } from './carousel-image';
import { uniqueId } from '../util/util';
let NglCarouselIndicator = class NglCarouselIndicator {
    constructor(el) {
        this.el = el;
        this.uid = uniqueId('carousel-indicator');
    }
    get tabindex() {
        return this.isActive ? 0 : -1;
    }
    ngOnChanges(changes) {
        this.image.active = this.isActive;
        if (changes.image) {
            this.image.labelledby = this.uid;
        }
    }
    focus() {
        this.el.nativeElement.focus();
    }
};
NglCarouselIndicator.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    HostBinding('class.slds-is-active'),
    HostBinding('attr.aria-selected'),
    Input(), InputBoolean(),
    __metadata("design:type", Object)
], NglCarouselIndicator.prototype, "isActive", void 0);
__decorate([
    HostBinding('attr.tabindex'),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [])
], NglCarouselIndicator.prototype, "tabindex", null);
__decorate([
    Input(),
    __metadata("design:type", NglCarouselImage)
], NglCarouselIndicator.prototype, "image", void 0);
__decorate([
    HostBinding('attr.id'),
    __metadata("design:type", Object)
], NglCarouselIndicator.prototype, "uid", void 0);
NglCarouselIndicator = __decorate([
    Directive({
        selector: '[nglCarouselIndicator]',
    }),
    __metadata("design:paramtypes", [ElementRef])
], NglCarouselIndicator);
export { NglCarouselIndicator };
